<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>epiwraps: Wrappers for plotting and dealing with epigenomics data • epiwraps</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="epiwraps: Wrappers for plotting and dealing with epigenomics data">
<meta name="description" content="A set of wrappers to facilitate working with epigenomics data. Includes in particular wrappers for producing multitrack single-regions plots, enrichment heatmaps, bigwig generation, normalization, etc. The focus was put on a simple yet flexible interface.">
<meta property="og:description" content="A set of wrappers to facilitate working with epigenomics data. Includes in particular wrappers for producing multitrack single-regions plots, enrichment heatmaps, bigwig generation, normalization, etc. The focus was put on a simple yet flexible interface.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">epiwraps</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.97</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/bam2bw.html">Generating bigwig tracks with bam2bw</a></li>
    <li><a class="dropdown-item" href="articles/misc.html">Miscellaneous epiwraps functions</a></li>
    <li><a class="dropdown-item" href="articles/multiRegionPlot.html">Visualizing signals across many regions</a></li>
    <li><a class="dropdown-item" href="articles/normalization.html">Normalizing genomic signals</a></li>
    <li><a class="dropdown-item" href="articles/singleRegionPlot.html">Visualizing signals in a single region</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="epiwraps-smooth-epigenomics-data-analysis-and-visualization-with-bioconductor">epiwraps: smooth epigenomics data analysis and visualization with Bioconductor<a class="anchor" aria-label="anchor" href="#epiwraps-smooth-epigenomics-data-analysis-and-visualization-with-bioconductor"></a>
</h1></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Bioconductor offers powerful packages for the analysis and visualization of (epi)genomic data, however they are often not always easy to approach for users without an extensive bioinformatics background. The <code>epiwraps</code> package started off as a set of wrappers for the visualization of epigenomics data (in particular ATAC/ChIP-seq), meant to facilitate the teaching of regulatory genomics through hands-on exploration of such data. It has now evolved into a set of tools around this task, including missing R-based alternatives for some important steps and a more unified interface.</p>
<p><strong>Note that the package and documentation are both still under heavy development!</strong></p>
<div class="section level3">
<h3 id="recent-changes">Recent changes<a class="anchor" aria-label="anchor" href="#recent-changes"></a>
</h3>
<p>Version &gt;0.99.80 introduced some major changes, in particular the introduction of <code>EnrichmentSE</code> objects, which are now the default output of <code>signal2Matrix</code> and facilitate manipulations, as well as renaming of some normalization functions under a more standardized interface. Please consult the vignette to update your workflows.</p>
</div>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>Install with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ETHZ-INS/epiwraps"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>While the package’s functions accept many input formats, much of the work usually done involving it uses bigwig files, and the package covers both the flexible <a href="https://ethz-ins.github.io/epiwraps/articles/bam2bw.html">generation</a> of such files, as well as their use for extracting and plotting data, whether in a <a href="https://ethz-ins.github.io/epiwraps/articles/singleRegionPlot.html">single region</a> or <a href="https://ethz-ins.github.io/epiwraps/articles/multiRegionPlot.html">across many regions</a>.</p>
<p>This is exemplified in the following figure:</p>
<p><img src="articles/example_bigwigs_hm.png" width="700"></p>
<p>All columns are based on the same bam file, but select and/or summarize fragments differently. For example, the heatmap of insertion sites of nucleosome-free fragments was based on a bigwig file generated using:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/bam2bw.html">bam2bw</a></span><span class="op">(</span><span class="st">"aligned.bam"</span>, output_bw<span class="op">=</span><span class="st">"NF_cuts.bw"</span>, paired<span class="op">=</span><span class="cn">TRUE</span>, <span class="va">`shift=c(4L,-5L), </span></span>
<span><span class="va">       binWidth=1L, minFragLength=25, maxFragLength=120, type="ends"`</span><span class="op">)</span></span></code></pre></div>
<p>More information about bigwig file generation is available in the <a href="https://ethz-ins.github.io/epiwraps/articles/bam2bw.html">bam2bw vignette</a>.</p>
<p>For information on the extraction of signal from such files (as well as other format) and plotting of such heatmaps, see the <a href="https://ethz-ins.github.io/epiwraps/articles/multiRegionPlot.html">vignette to this effect</a>.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing epiwraps</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Pierre-Luc Germain <br><small class="roles"> Maintainer, author </small> <a href="https://orcid.org/0000-0003-3418-4218" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Emanuel Sonder <br><small class="roles"> Author </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre-Luc Germain, Emanuel Sonder.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
