<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="bam2bw: create a coverage bigwig file from an alignment bam file."><title>bam2bw: create a coverage bigwig file from an alignment bam file. — bam2bw • epiwraps</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="bam2bw: create a coverage bigwig file from an alignment bam file. — bam2bw"><meta property="og:description" content="bam2bw: create a coverage bigwig file from an alignment bam file."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">epiwraps</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.94</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/bam2bw.html">Generating bigwig tracks with bam2bw</a>
    <a class="dropdown-item" href="../articles/misc.html">Miscellaneous epiwraps functions</a>
    <a class="dropdown-item" href="../articles/multiRegionPlot.html">Visualizing signals across many regions</a>
    <a class="dropdown-item" href="../articles/normalization.html">Normalizing genomic signals</a>
    <a class="dropdown-item" href="../articles/singleRegionPlot.html">Visualizing signals in a single region</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>bam2bw: create a coverage bigwig file from an alignment bam file.</h1>
      
      <div class="d-none name"><code>bam2bw.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>bam2bw: create a coverage bigwig file from an alignment bam file.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">bam2bw</span><span class="op">(</span></span>
<span>  <span class="va">bamfile</span>,</span>
<span>  <span class="va">output_bw</span>,</span>
<span>  bgbam <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  paired <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  binWidth <span class="op">=</span> <span class="fl">20L</span>,</span>
<span>  extend <span class="op">=</span> <span class="fl">0L</span>,</span>
<span>  scaling <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  shift <span class="op">=</span> <span class="fl">0L</span>,</span>
<span>  type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"full"</span>, <span class="st">"center"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"ends"</span><span class="op">)</span>,</span>
<span>  compType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"log2ratio"</span>, <span class="st">"subtract"</span>, <span class="st">"log10ppois"</span>, <span class="st">"log10FE"</span><span class="op">)</span>,</span>
<span>  strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"*"</span>, <span class="st">"+"</span>, <span class="st">"-"</span><span class="op">)</span>,</span>
<span>  strandMode <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  log1p <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  exclude <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  includeDuplicates <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  includeSecondary <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  minMapq <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  minFragLength <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  maxFragLength <span class="op">=</span> <span class="fl">5000L</span>,</span>
<span>  keepSeqLvls <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  splitByChr <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  pseudocount <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  localBackground <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  only <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  zeroCap <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  forceSeqlevelsStyle <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  binSummarization <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mean"</span>, <span class="st">"max"</span>, <span class="st">"min"</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>bamfile</dt>
<dd><p>The path to the signal bam file.</p></dd>


<dt>output_bw</dt>
<dd><p>The path to the output bigwig file</p></dd>


<dt>bgbam</dt>
<dd><p>The optional path to the control/input bam file to compute 
relative signal (see the `compType` argument).</p></dd>


<dt>paired</dt>
<dd><p>Logical; whether to consider fragments instead of reads for 
paired-end data. For spliced (e.g. RNA-seq) data, paired should always be 
set to FALSE.</p></dd>


<dt>binWidth</dt>
<dd><p>The window size. A lower value (min 1) means a higher 
resolution, but larger file size.</p></dd>


<dt>extend</dt>
<dd><p>The amount *by* which to extend single-end reads (e.g. 
fragment length minus read length). If `paired=TRUE` and `type` is either
'ends' or 'center', then the extension will be applied after taking the 
(shifted) fragment ends or centers, resulting in ranges of width equal to 
`extend`.</p></dd>


<dt>scaling</dt>
<dd><p>Either TRUE (performs Count Per Million scaling), FALSE (no 
scaling), or a numeric value by which the signal will be divided. If 
`bgbam` is given and `scaling=TRUE`, the background will be scaled to the
main signal.</p></dd>


<dt>shift</dt>
<dd><p>Shift (from 3' to 5') by which reads/fragments will be shifted.
If `shift` is an integer vector of length 2, the first value will represent
the shift for the positive strand, and the second for the negative strand.
For ATACseq data, one normally uses `shift=c(4L,-5L)`.</p></dd>


<dt>type</dt>
<dd><p>Type of the coverage to compile. Either full (full read/fragment),
start (count read/fragment start locations), end, center, or 'ends' (both
ends of the read/fragment).</p></dd>


<dt>compType</dt>
<dd><p>The type of relative signal to compute (ignored if `bgbam` 
isn't given). Can either be 'log2ratio' (log2-foldchange between signals),
'subtract' (difference), 'log10ppois' (rounded -log10 poisson p-value),
'log10FE' (log10 fold-enrichment). For fold-based signals, `pseudocount` 
will be added before computing the ratio. By default negative values are
capped (see `zeroCap`).</p></dd>


<dt>strand</dt>
<dd><p>Strand(s) to capture (any by default).</p></dd>


<dt>strandMode</dt>
<dd><p>The strandMode of the data (whether the strand is given by
the first or second mate, which depends on the library prep protocol). See
<a href="https://rdrr.io/pkg/GenomicAlignments/man/GAlignmentPairs-class.html" class="external-link">strandMode</a> for more information. This parameter 
has no effect unless one of the `strand`, `extend` parameters or a 
strand-specific `shift` are used.</p></dd>


<dt>log1p</dt>
<dd><p>Whether to log-transform (`log(x+1)`) the (scaled) signal.
Ignored when the signal is relative to a background.</p></dd>


<dt>exclude</dt>
<dd><p>An optional GRanges of regions for which overlapping reads 
should be excluded.</p></dd>


<dt>includeDuplicates</dt>
<dd><p>Logical, whether to include reads flagged as 
duplicates.</p></dd>


<dt>includeSecondary</dt>
<dd><p>Logical; whether to include secondary alignments</p></dd>


<dt>minMapq</dt>
<dd><p>Minimum mapping quality (1 to 255)</p></dd>


<dt>minFragLength</dt>
<dd><p>Minimum fragment length (ignored if `paired=FALSE`)</p></dd>


<dt>maxFragLength</dt>
<dd><p>Maximum fragment length (ignored if `paired=FALSE`)</p></dd>


<dt>keepSeqLvls</dt>
<dd><p>An optional vector of seqLevels (i.e. chromosomes) to 
include.</p></dd>


<dt>splitByChr</dt>
<dd><p>Whether to process chromosomes separately, and if so by how
many chunks. The should not affect the output, and is simply slightly 
slower and consumes less memory. Can be a logical value (in which case each
chromosome is processed separately), but we instead recommend giving a 
positive integer indicating the number of chunks.</p></dd>


<dt>pseudocount</dt>
<dd><p>The count to be added before fold-enrichment calculation 
between signals. Ignored if `compType="subtract"`.</p></dd>


<dt>localBackground</dt>
<dd><p>A (vector of) number of windows around each 
tile/position for which a local background will be calculated. The 
background used will be the maximum of the one at the given position or of
the mean of each of the local backgrounds.</p></dd>


<dt>only</dt>
<dd><p>An optional GRanges of regions for which overlapping reads should
be included. If set, all other reads are discarded.</p></dd>


<dt>zeroCap</dt>
<dd><p>Logical; whether to cap values below zero to zero when 
producing relative signals.</p></dd>


<dt>forceSeqlevelsStyle</dt>
<dd><p>If specified, forces the use of the specified 
seqlevel style for the output bigwig. Can take any value accepted by
`seqlevelsStyle`.</p></dd>


<dt>verbose</dt>
<dd><p>Logical; whether to print progress messages</p></dd>


<dt>binSummarization</dt>
<dd><p>The method to summarize nucleotides into each bin,
either "max" (default), "min" or "mean".</p></dd>


<dt>...</dt>
<dd><p>Passed to `ScanBamParam`</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>The bigwig filepath. Alternatively, if `output_bw=NA_character_`, 
  the coverage data is not written to file but returned.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>* For single-end ChIPseq data, extend reads to the expected fragment size 
  using the `extend` argument.
* The implementation involves reading the reads before processing, which can
 be quite memory-hungry. To avoid this the files are read by chunks (composed
  of chromosomes of balanced sizes), controlled by the `splitByChr` argument.
  This reduces memory usage but also creates overhead which reduces the 
  speed. In contexts where the file is small or memory isn't a problem, using 
 `splitByChr=FALSE` will achieve the best speed. Otherwise, increasing 
 `splitByChr` will decrease the memory footprint.
* Consider restricting the read quality using `includeDuplicates=FALSE` 
  and `minMapq=20` for example.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Pierre-Luc Germain</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># get an example bam file</span></span></span>
<span class="r-in"><span><span class="va">bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"ex1.bam"</span>, package<span class="op">=</span><span class="st">"Rsamtools"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># create bigwig</span></span></span>
<span class="r-in"><span><span class="fu">bam2bw</span><span class="op">(</span><span class="va">bam</span>, <span class="st">"output.bw"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `paired` not specified, assuming single-end reads. Set to paired='auto' to automatically detect.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading in signal...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Writing bigwig...</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre-Luc Germain, Emanuel Sonder.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

