<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>callPeaks — callPeaks • epiwraps</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="callPeaks — callPeaks"><meta name="description" content="This is a R-based implementation of the general MACS2 strategy (Zhang et al.,
 Genome Biology 2008), taking some freedom here and there in comparison to
 the original. The function is still under development, especially with
 respect to single-end reads, where some optimization might still be needed.
 For paired-end reads, the results are nearly identical with those of MACS2,
 with two main differences: 1) the p-values are more conservative (and
 arguably better calibrated) and 2) because the implementation does not rely
 on sliding windows, with default settings the peaks are narrower."><meta property="og:description" content="This is a R-based implementation of the general MACS2 strategy (Zhang et al.,
 Genome Biology 2008), taking some freedom here and there in comparison to
 the original. The function is still under development, especially with
 respect to single-end reads, where some optimization might still be needed.
 For paired-end reads, the results are nearly identical with those of MACS2,
 with two main differences: 1) the p-values are more conservative (and
 arguably better calibrated) and 2) because the implementation does not rely
 on sliding windows, with default settings the peaks are narrower."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">epiwraps</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.96</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/bam2bw.html">Generating bigwig tracks with bam2bw</a></li>
    <li><a class="dropdown-item" href="../articles/misc.html">Miscellaneous epiwraps functions</a></li>
    <li><a class="dropdown-item" href="../articles/multiRegionPlot.html">Visualizing signals across many regions</a></li>
    <li><a class="dropdown-item" href="../articles/normalization.html">Normalizing genomic signals</a></li>
    <li><a class="dropdown-item" href="../articles/singleRegionPlot.html">Visualizing signals in a single region</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>callPeaks</h1>

      <div class="d-none name"><code>callPeaks.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This is a R-based implementation of the general MACS2 strategy (Zhang et al.,
 Genome Biology 2008), taking some freedom here and there in comparison to
 the original. The function is still under development, especially with
 respect to single-end reads, where some optimization might still be needed.
 For paired-end reads, the results are nearly identical with those of MACS2,
 with two main differences: 1) the p-values are more conservative (and
 arguably better calibrated) and 2) because the implementation does not rely
 on sliding windows, with default settings the peaks are narrower.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">callPeaks</span><span class="op">(</span></span>
<span>  <span class="va">bam</span>,</span>
<span>  ctrl <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  paired <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"narrow"</span>, <span class="st">"broad"</span><span class="op">)</span>,</span>
<span>  nullH <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"local"</span>, <span class="st">"global.nb"</span>, <span class="st">"global.bin"</span><span class="op">)</span>,</span>
<span>  blacklist <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  binSize <span class="op">=</span> <span class="fl">10L</span>,</span>
<span>  fragLength <span class="op">=</span> <span class="fl">200L</span>,</span>
<span>  minPeakCount <span class="op">=</span> <span class="fl">5L</span>,</span>
<span>  minFoldChange <span class="op">=</span> <span class="fl">1.3</span>,</span>
<span>  pthres <span class="op">=</span> <span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">3</span>,</span>
<span>  maxSize <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bgWindow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">10</span><span class="op">)</span> <span class="op">*</span> <span class="fl">1000</span>,</span>
<span>  pseudoCount <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  useStrand <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  outFormat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"custom"</span>, <span class="st">"narrowPeak"</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-bam">bam<a class="anchor" aria-label="anchor" href="#arg-bam"></a></dt>
<dd><p>A signal bam file</p></dd>


<dt id="arg-ctrl">ctrl<a class="anchor" aria-label="anchor" href="#arg-ctrl"></a></dt>
<dd><p>An optional (but highly recommended) control bam file</p></dd>


<dt id="arg-paired">paired<a class="anchor" aria-label="anchor" href="#arg-paired"></a></dt>
<dd><p>Logical, whether the reads are paired</p></dd>


<dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>The type of peaks to identify ('narrow' or 'broad').</p></dd>


<dt id="arg-blacklist">blacklist<a class="anchor" aria-label="anchor" href="#arg-blacklist"></a></dt>
<dd><p>An optional `GRanges` of regions to be excluded (or the path
to such a file). Since the blacklisted regions are removed from both the
signal and control peaks, this also has an important impact on the
empirical FDR (when `ctrl` is given).</p></dd>


<dt id="arg-binsize">binSize<a class="anchor" aria-label="anchor" href="#arg-binsize"></a></dt>
<dd><p>Binsize used to estimate peak shift</p></dd>


<dt id="arg-fraglength">fragLength<a class="anchor" aria-label="anchor" href="#arg-fraglength"></a></dt>
<dd><p>Fragment length. Ignored if `paired=TRUE`. This is only
used for the initial candidate region identification, and sizes are
adjusted after, so it doesn't need to be very precise.</p></dd>


<dt id="arg-minpeakcount">minPeakCount<a class="anchor" aria-label="anchor" href="#arg-minpeakcount"></a></dt>
<dd><p>The minimum summit count for a region to be considered.
Decreasing this can substantially increase the running time.</p></dd>


<dt id="arg-minfoldchange">minFoldChange<a class="anchor" aria-label="anchor" href="#arg-minfoldchange"></a></dt>
<dd><p>The minimum fold-change for a region to be considered.
Decreasing this can substantially increase the running time.</p></dd>


<dt id="arg-pthres">pthres<a class="anchor" aria-label="anchor" href="#arg-pthres"></a></dt>
<dd><p>The p-value threshold to use</p></dd>


<dt id="arg-bgwindow">bgWindow<a class="anchor" aria-label="anchor" href="#arg-bgwindow"></a></dt>
<dd><p>The windows to consider (in addition to the peak itself)
for local background.</p></dd>


<dt id="arg-outformat">outFormat<a class="anchor" aria-label="anchor" href="#arg-outformat"></a></dt>
<dd><p>The output format ('custom' or 'narrowPeak')</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Logical; whether to output progress messages</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Passed to <code><a href="bamChrChunkApply.html">bamChrChunkApply</a></code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A `GRanges`</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>`callPeaks` takes about twice as long to run as MACS2, and uses more memory.
If dealing with very large files (or a very low memory system), consider
increasing the number of processing chunks, for instance with `nChunks=10`.</p>
<p>The function uses <code><a href="bamChrChunkApply.html">bamChrChunkApply</a></code> to obtain the coverages,
and can accept any argument of that function. This means that the
`mapqFilter` and bam `flgs` arguments can be used to restrict the reads used.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre-Luc Germain, Emanuel Sonder.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

