<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>frag2bw — frag2bw • epiwraps</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="frag2bw — frag2bw"><meta name="description" content="Creates a coverage bigwig file from a Tabix-indexed fragment file."><meta property="og:description" content="Creates a coverage bigwig file from a Tabix-indexed fragment file."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">epiwraps</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.105</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/bam2bw.html">Generating bigwig tracks with bam2bw</a></li>
    <li><a class="dropdown-item" href="../articles/misc.html">Miscellaneous epiwraps functions</a></li>
    <li><a class="dropdown-item" href="../articles/multiRegionPlot.html">Visualizing signals across many regions</a></li>
    <li><a class="dropdown-item" href="../articles/normalization.html">Normalizing genomic signals</a></li>
    <li><a class="dropdown-item" href="../articles/singleRegionPlot.html">Visualizing signals in a single region</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>frag2bw</h1>

      <div class="d-none name"><code>frag2bw.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Creates a coverage bigwig file from a Tabix-indexed fragment file.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">frag2bw</span><span class="op">(</span></span>
<span>  <span class="va">tabixFile</span>,</span>
<span>  <span class="va">output_bw</span>,</span>
<span>  paired <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  binWidth <span class="op">=</span> <span class="fl">20L</span>,</span>
<span>  extend <span class="op">=</span> <span class="fl">0L</span>,</span>
<span>  scaling <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"full"</span>, <span class="st">"center"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"ends"</span><span class="op">)</span>,</span>
<span>  barcodes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"*"</span>, <span class="st">"+"</span>, <span class="st">"-"</span><span class="op">)</span>,</span>
<span>  shift <span class="op">=</span> <span class="fl">0L</span>,</span>
<span>  log1p <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  exclude <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minFragLength <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  maxFragLength <span class="op">=</span> <span class="fl">5000L</span>,</span>
<span>  keepSeqLvls <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  useScore <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  forceSeqlevelsStyle <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  only <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  format <span class="op">=</span> <span class="st">"bed"</span>,</span>
<span>  binSummarization <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"max"</span>, <span class="st">"min"</span>, <span class="st">"mean"</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-tabixfile">tabixFile<a class="anchor" aria-label="anchor" href="#arg-tabixfile"></a></dt>
<dd><p>The path to a tabix-indexed bam file, or a TabixFile object.</p></dd>


<dt id="arg-output-bw">output_bw<a class="anchor" aria-label="anchor" href="#arg-output-bw"></a></dt>
<dd><p>The path to the output bigwig file</p></dd>


<dt id="arg-paired">paired<a class="anchor" aria-label="anchor" href="#arg-paired"></a></dt>
<dd><p>Logical; whether the coordinates are that of fragments, as
opposed to single-end reads where the only one end of the fragments is given.
TRUE by default.</p></dd>


<dt id="arg-binwidth">binWidth<a class="anchor" aria-label="anchor" href="#arg-binwidth"></a></dt>
<dd><p>The window size. A lower value (min 1) means a higher
resolution, but larger file size.</p></dd>


<dt id="arg-scaling">scaling<a class="anchor" aria-label="anchor" href="#arg-scaling"></a></dt>
<dd><p>Either TRUE (performs Count Per Million scaling), FALSE (no
scaling), or a numeric value by which the signal will be divided. If
`bgbam` is given and `scaling=TRUE`, the background will be scaled to the
main signal.</p></dd>


<dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>Type of the coverage to compile. Either full (full read/fragment),
start (count read/fragment start locations), end, center, or 'ends' (both
ends of the read/fragment).</p></dd>


<dt id="arg-barcodes">barcodes<a class="anchor" aria-label="anchor" href="#arg-barcodes"></a></dt>
<dd><p>An optional list of barcodes to use (assuming that the file
contains the column)</p></dd>


<dt id="arg-strand">strand<a class="anchor" aria-label="anchor" href="#arg-strand"></a></dt>
<dd><p>Strand(s) to capture (any by default).</p></dd>


<dt id="arg-shift">shift<a class="anchor" aria-label="anchor" href="#arg-shift"></a></dt>
<dd><p>Shift (from 3' to 5') by which reads/fragments will be shifted.
If `shift` is an integer vector of length 2, the first value will represent
the shift for the positive strand, and the second for the negative strand.</p></dd>


<dt id="arg-exclude">exclude<a class="anchor" aria-label="anchor" href="#arg-exclude"></a></dt>
<dd><p>An optional GRanges of regions for which overlapping reads
should be excluded.</p></dd>


<dt id="arg-minfraglength">minFragLength<a class="anchor" aria-label="anchor" href="#arg-minfraglength"></a></dt>
<dd><p>Minimum fragment length (ignored if `paired=FALSE`)</p></dd>


<dt id="arg-maxfraglength">maxFragLength<a class="anchor" aria-label="anchor" href="#arg-maxfraglength"></a></dt>
<dd><p>Maximum fragment length (ignored if `paired=FALSE`)</p></dd>


<dt id="arg-keepseqlvls">keepSeqLvls<a class="anchor" aria-label="anchor" href="#arg-keepseqlvls"></a></dt>
<dd><p>An optional vector of seqLevels (i.e. chromosomes) to
include.</p></dd>


<dt id="arg-usescore">useScore<a class="anchor" aria-label="anchor" href="#arg-usescore"></a></dt>
<dd><p>Whether to use the score column (if any) as coverage weights.</p></dd>


<dt id="arg-forceseqlevelsstyle">forceSeqlevelsStyle<a class="anchor" aria-label="anchor" href="#arg-forceseqlevelsstyle"></a></dt>
<dd><p>If specified, forces the use of the specified
seqlevel style for the output bigwig. Can take any value accepted by
`seqlevelsStyle`.</p></dd>


<dt id="arg-only">only<a class="anchor" aria-label="anchor" href="#arg-only"></a></dt>
<dd><p>An optional GRanges of regions for which overlapping reads should
be included. If set, all other reads are discarded.</p></dd>


<dt id="arg-format">format<a class="anchor" aria-label="anchor" href="#arg-format"></a></dt>
<dd><p>The format of the fragment file.</p></dd>


<dt id="arg-binsummarization">binSummarization<a class="anchor" aria-label="anchor" href="#arg-binsummarization"></a></dt>
<dd><p>The method to summarize nucleotides into each bin,
either "max" (default), "min" or "mean".</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Logical; whether to print progress messages</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>The bigwig filepath. Alternatively, if `output_bw=NA_character_`,
  the coverage data is not written to file but returned.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># we first create a fake tabix file:</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">reads</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1"</span>,<span class="st">"2"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                 <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">5000</span><span class="op">+</span><span class="fl">10</span><span class="op">*</span><span class="fl">1</span><span class="op">:</span><span class="fl">7</span>, width<span class="op">=</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bedf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext<span class="op">=</span><span class="st">".bed"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/BEDFile-class.html" class="external-link">export.bed</a></span><span class="op">(</span><span class="va">reads</span>, <span class="va">bedf</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bedf</span> <span class="op">&lt;-</span> <span class="fu">Rsamtools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/zip.html" class="external-link">bgzip</a></span><span class="op">(</span><span class="va">bedf</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">Rsamtools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/indexTabix.html" class="external-link">indexTabix</a></span><span class="op">(</span><span class="va">bedf</span>, format<span class="op">=</span><span class="st">"bed"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "/tmp/Rtmp51bmUt/file16bd2b929504.bed.bgz.tbi"</span>
<span class="r-in"><span><span class="co"># convert to bigwig</span></span></span>
<span class="r-in"><span><span class="fu">frag2bw</span><span class="op">(</span><span class="va">bedf</span>, <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext<span class="op">=</span><span class="st">".bw"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading in signal...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Writing bigwig...</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre-Luc Germain, Emanuel Sonder.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

