<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>getCounts — getCounts • epiwraps</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="getCounts — getCounts"><meta name="description" content="Creates a SummarizedExperiment of fragment (or insertion) counts from bam
files that overlap given regions."><meta property="og:description" content="Creates a SummarizedExperiment of fragment (or insertion) counts from bam
files that overlap given regions."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">epiwraps</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.107</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/bam2bw.html">Generating bigwig tracks with bam2bw</a></li>
    <li><a class="dropdown-item" href="../articles/misc.html">Miscellaneous epiwraps functions</a></li>
    <li><a class="dropdown-item" href="../articles/multiRegionPlot.html">Visualizing signals across many regions</a></li>
    <li><a class="dropdown-item" href="../articles/normalization.html">Normalizing genomic signals</a></li>
    <li><a class="dropdown-item" href="../articles/singleRegionPlot.html">Visualizing signals in a single region</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>getCounts</h1>

      <div class="d-none name"><code>getCounts.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Creates a SummarizedExperiment of fragment (or insertion) counts from bam
files that overlap given regions.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">getCounts</span><span class="op">(</span></span>
<span>  <span class="va">bam_files</span>,</span>
<span>  <span class="va">regions</span>,</span>
<span>  <span class="va">paired</span>,</span>
<span>  extend <span class="op">=</span> <span class="fl">0L</span>,</span>
<span>  shift <span class="op">=</span> <span class="fl">0L</span>,</span>
<span>  type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"full"</span>, <span class="st">"center"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"ends"</span><span class="op">)</span>,</span>
<span>  ov.type <span class="op">=</span> <span class="st">"any"</span>,</span>
<span>  maxgap <span class="op">=</span> <span class="op">-</span><span class="fl">1L</span>,</span>
<span>  minoverlap <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  ignore.strand <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  strandMode <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  includeDuplicates <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  includeSecondary <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  minMapq <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  minFragLength <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  maxFragLength <span class="op">=</span> <span class="fl">5000L</span>,</span>
<span>  splitByChr <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  randomAcc <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-bam-files">bam_files<a class="anchor" aria-label="anchor" href="#arg-bam-files"></a></dt>
<dd><p>A vector of paths to the bam files.</p></dd>


<dt id="arg-regions">regions<a class="anchor" aria-label="anchor" href="#arg-regions"></a></dt>
<dd><p>A `GRanges` of regions in which to counts.</p></dd>


<dt id="arg-paired">paired<a class="anchor" aria-label="anchor" href="#arg-paired"></a></dt>
<dd><p>Logical; whether the data is paired (assumed unpaired by
default). Use `paired="auto"` for automatic detection using the first bam
file.</p></dd>


<dt id="arg-extend">extend<a class="anchor" aria-label="anchor" href="#arg-extend"></a></dt>
<dd><p>The amount *by* which to extend single-end reads (e.g.
fragment length minus read length). If `paired=TRUE` and `type` is either
'ends' or 'center', then the extension will be applied after taking the
(shifted) fragment ends or centers, resulting in ranges of width equal to
`extend`.</p></dd>


<dt id="arg-shift">shift<a class="anchor" aria-label="anchor" href="#arg-shift"></a></dt>
<dd><p>Shift (from 3' to 5') by which reads/fragments will be shifted.
If `shift` is an integer vector of length 2, the first value will represent
the shift for the positive strand, and the second for the negative strand.</p></dd>


<dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>Type of the coverage to compile. Either full (full read/fragment),
start (count read/fragment start locations), end, center, or 'ends' (both
ends of the read/fragment).</p></dd>


<dt id="arg-ov-type">ov.type<a class="anchor" aria-label="anchor" href="#arg-ov-type"></a></dt>
<dd><p>Overlap type. See the `type` argument of
<code>link[GenomicRanges]{countOverlaps}</code>.</p></dd>


<dt id="arg-maxgap">maxgap<a class="anchor" aria-label="anchor" href="#arg-maxgap"></a></dt>
<dd><p>Maximum gap allowed for overlaps (see the corresponding
argument of <code>link[GenomicRanges]{countOverlaps}</code>).</p></dd>


<dt id="arg-minoverlap">minoverlap<a class="anchor" aria-label="anchor" href="#arg-minoverlap"></a></dt>
<dd><p>Minimum overlap (see the corresponding argument of
<code>link[GenomicRanges]{countOverlaps}</code>).</p></dd>


<dt id="arg-ignore-strand">ignore.strand<a class="anchor" aria-label="anchor" href="#arg-ignore-strand"></a></dt>
<dd><p>Logical; whether to ignore strand for the purpose of
counting overlaps (default TRUE).</p></dd>


<dt id="arg-strandmode">strandMode<a class="anchor" aria-label="anchor" href="#arg-strandmode"></a></dt>
<dd><p>The strandMode of the data (whether the strand is given by
the first or second mate, which depends on the library prep protocol). See
<a href="https://rdrr.io/pkg/GenomicAlignments/man/GAlignmentPairs-class.html" class="external-link">strandMode</a> for more information. This parameter
has no effect unless one of the `strand`, `extend` parameters or a
strand-specific `shift` are used.</p></dd>


<dt id="arg-includeduplicates">includeDuplicates<a class="anchor" aria-label="anchor" href="#arg-includeduplicates"></a></dt>
<dd><p>Logical, whether to include reads flagged as
duplicates.</p></dd>


<dt id="arg-includesecondary">includeSecondary<a class="anchor" aria-label="anchor" href="#arg-includesecondary"></a></dt>
<dd><p>Logical; whether to include secondary alignments</p></dd>


<dt id="arg-minmapq">minMapq<a class="anchor" aria-label="anchor" href="#arg-minmapq"></a></dt>
<dd><p>Minimum mapping quality (1 to 255)</p></dd>


<dt id="arg-minfraglength">minFragLength<a class="anchor" aria-label="anchor" href="#arg-minfraglength"></a></dt>
<dd><p>Minimum fragment length (ignored if `paired=FALSE`)</p></dd>


<dt id="arg-maxfraglength">maxFragLength<a class="anchor" aria-label="anchor" href="#arg-maxfraglength"></a></dt>
<dd><p>Maximum fragment length (ignored if `paired=FALSE`)</p></dd>


<dt id="arg-splitbychr">splitByChr<a class="anchor" aria-label="anchor" href="#arg-splitbychr"></a></dt>
<dd><p>Whether to process chromosomes separately, and if so by how
many chunks. The should not affect the output, and is simply slightly
slower and consumes less memory. Can be a logical value (in which case each
chromosome is processed separately), but we instead recommend giving a
positive integer indicating the number of chunks.</p></dd>


<dt id="arg-randomacc">randomAcc<a class="anchor" aria-label="anchor" href="#arg-randomacc"></a></dt>
<dd><p>Logical, whether to use random access. This is disabled by
default because the overhead of random access to a lot of regions is
typically worse than reading the entire file. However, if you need to get
counts in few regions, enabling this will be faster. Note however that
when using random access, the output object will not contain depth
information.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Logical; whether to print progress messages</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A RangedSummarizedExperiment with a 'counts' assay.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># get an example bam file</span></span></span>
<span class="r-in"><span><span class="va">bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"ex1.bam"</span>, package<span class="op">=</span><span class="st">"Rsamtools"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># create regions of interest</span></span></span>
<span class="r-in"><span><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seq1"</span>,<span class="st">"seq1"</span>,<span class="st">"seq2"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">400</span>,<span class="fl">900</span>,<span class="fl">500</span><span class="op">)</span>, width<span class="op">=</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">getCounts</span><span class="op">(</span><span class="va">bam</span>, <span class="va">peaks</span>, paired<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in h(simpleError(msg, call)):</span> error in evaluating the argument 'x' in selecting a method for function 'seqlevelsInUse': object 'peaks' not found</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre-Luc Germain, Emanuel Sonder.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

