<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>getNormFactors : estimate normalization factors from genomic signal files — getNormFactors • epiwraps</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="getNormFactors : estimate normalization factors from genomic signal files — getNormFactors"><meta name="description" content="Estimates normalization factors for a set of samples (i.e. bam/bigwig files)."><meta property="og:description" content="Estimates normalization factors for a set of samples (i.e. bam/bigwig files)."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">epiwraps</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.103</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/bam2bw.html">Generating bigwig tracks with bam2bw</a></li>
    <li><a class="dropdown-item" href="../articles/misc.html">Miscellaneous epiwraps functions</a></li>
    <li><a class="dropdown-item" href="../articles/multiRegionPlot.html">Visualizing signals across many regions</a></li>
    <li><a class="dropdown-item" href="../articles/normalization.html">Normalizing genomic signals</a></li>
    <li><a class="dropdown-item" href="../articles/singleRegionPlot.html">Visualizing signals in a single region</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>getNormFactors : estimate normalization factors from genomic signal files</h1>

      <div class="d-none name"><code>getNormFactors.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Estimates normalization factors for a set of samples (i.e. bam/bigwig files).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">getNormFactors</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"background"</span>, <span class="st">"SES"</span>, <span class="st">"enriched"</span>, <span class="st">"top"</span>, <span class="st">"MAnorm"</span>, <span class="st">"S3norm"</span>, <span class="st">"2cLinear"</span><span class="op">)</span>,</span>
<span>  wsize <span class="op">=</span> <span class="fl">10L</span>,</span>
<span>  nwind <span class="op">=</span> <span class="fl">20000L</span>,</span>
<span>  peaks <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  trim <span class="op">=</span> <span class="fl">0.02</span>,</span>
<span>  useSeqLevels <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  paired <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">bwNormFactors</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A vector of paths to bigwig files, to bam files, or alternatively a
list of coverages in RleList format. (Mixed formats are not supported)</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>Normalization method (see details below).</p></dd>


<dt id="arg-wsize">wsize<a class="anchor" aria-label="anchor" href="#arg-wsize"></a></dt>
<dd><p>The size of the random windows. If any of the bigwig files
records point events (e.g. insertions) at high resolution (e.g. nucleotide),
use a lot (e.g. &gt;10k) of small windows (e.g. `wsize=10`), as per default
settings. Otherwise the process can be lightened by using fewer bigger
windows.</p></dd>


<dt id="arg-nwind">nwind<a class="anchor" aria-label="anchor" href="#arg-nwind"></a></dt>
<dd><p>The number of random windows</p></dd>


<dt id="arg-peaks">peaks<a class="anchor" aria-label="anchor" href="#arg-peaks"></a></dt>
<dd><p>A list of peaks (GRanges) for each experiment in `x`, or a
vector of paths to such files, or a single GRanges of unified peaks to use
(e.g. for top/MAnorm).</p></dd>


<dt id="arg-trim">trim<a class="anchor" aria-label="anchor" href="#arg-trim"></a></dt>
<dd><p>Amount of trimming when calculating means.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A vector of normalization factors, or for the 'S3norm' and '2cLinear'
  methods, a numeric matrix with a number of rows equal to the length
  of `x`, and two columns indicating the alpha and beta terms.</p>
<p>A vector of normalization factors or, for methods 'S3norm' and
  '2cLinear', a matrix of per-sample normalization parameters.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The function computes per-sample (bigwig or bam file) normalization factors
using one of the following methods:
* The 'background' or 'SES' normalization method (they are synonyms here)
(Diaz et al., Stat Appl Gen et Mol Biol, 2012) assumes that the background
noise should on average be the same across experiments, an assumption that
works well in practice when there are not very large differences in
signal-to-noise ratio. The implementation uses the trimmed mean number of
reads in random windows with non-zero counts.
* The 'MAnorm' approach (Shao et al., Genome Biology 2012) assumes that
regions that are commonly enriched (i.e. common peaks) in two experiments
should on average have the same signal in the two experiments. If the data is
strictly positive (e.g. counts or CPMs), TMM normalization (Robinson et al.,
Genome Biology 2010) is then employed on the common peaks. When done on more
than two samples, the sample with the highest number of overlaps with other
samples is used as a reference.
* The 'enriched' approach assumes that enriched regions are on average
similarly enriched across samples. Contrarily to 'MAnorm', these regions do
not need to be in common across samples/experiments. Note that trimming via
the `trim` parameter is performed before calculating means.
* The 'top' approach assumes that the maximum enrichment (after trimming
according to the `trim` parameter) in peaks is the same across
samples/experiments.
* The 'S3norm' (Xiang et al., NAR 2020) and '2cLinear' methods try to
normalize both enrichment and background simultaneously. S3norm does this in
a log-linear fashion (as in the publication), while '2cLinear' does it on
the original scale.</p>
<p>Several of the methods rely on random regions to estimate background levels.
This means that they will not be entirely deterministic, although normally
quite reproducible with experiments that are not very sparse. For fully
reproducible results, be sure to set the random seed.</p>
<p>When the data is very sparse (e.g. low sequencing depth, or compiling single
nucleotide hits rather than fragment coverage), it might be necessary to
increase the `wsize` and `nwind` to get more robust estimates.</p>
    </div>
    <div class="section level2">
    <h2 id="functions">Functions<a class="anchor" aria-label="anchor" href="#functions"></a></h2>

<ul><li><p><code>bwNormFactors()</code>: deprecated in favor of getNormFactors</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># we get an example bigwig file, and use it twice:</span></span></span>
<span class="r-in"><span><span class="va">bw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example_atac.bw"</span>, package<span class="op">=</span><span class="st">"epiwraps"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>sample1<span class="op">=</span><span class="va">bw</span>, sample2<span class="op">=</span><span class="va">bw</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># we indicate to use only chr1, because it's the only one in the example file</span></span></span>
<span class="r-in"><span><span class="fu">getNormFactors</span><span class="op">(</span><span class="va">bw</span>, useSeqLevels<span class="op">=</span><span class="st">"1"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Comparing coverage in random regions...</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Some samples have less than 50 non-zero windows. Consider increasing the window size or (better) the number of windows.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> sample1 sample2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       1       1 </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre-Luc Germain, Emanuel Sonder.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

