<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Estimates normalization factors for a set of samples (i.e. bam/bigwig files)."><title>getNormFactors : estimate normalization factors from genomic signal files — getNormFactors • epiwraps</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="getNormFactors : estimate normalization factors from genomic signal files — getNormFactors"><meta property="og:description" content="Estimates normalization factors for a set of samples (i.e. bam/bigwig files)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">epiwraps</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.85</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/bam2bw.html">Generating bigwig tracks with bam2bw</a>
    <a class="dropdown-item" href="../articles/misc.html">Miscellaneous epiwraps functions</a>
    <a class="dropdown-item" href="../articles/multiRegionPlot.html">Visualizing signals across many regions</a>
    <a class="dropdown-item" href="../articles/singleRegionPlot.html">Visualizing signals in a single region</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>getNormFactors : estimate normalization factors from genomic signal files</h1>
      
      <div class="d-none name"><code>getNormFactors.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Estimates normalization factors for a set of samples (i.e. bam/bigwig files).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">getNormFactors</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  wsize <span class="op">=</span> <span class="fl">10L</span>,</span>
<span>  nwind <span class="op">=</span> <span class="fl">20000L</span>,</span>
<span>  peaks <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  trim <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  useSeqLevels <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  paired <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"background"</span>, <span class="st">"SES"</span>, <span class="st">"MAnorm"</span>, <span class="st">"S3norm"</span>, <span class="st">"2cLinear"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">bwNormFactors</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>x</dt>
<dd><p>A vector of paths to bigwig files, to bam files, or alternatively a 
list of coverages in RleList format. (Mixed formats are not supported)</p></dd>


<dt>wsize</dt>
<dd><p>The size of the random windows. If any of the bigwig files 
records point events (e.g. insertions) at high resolution (e.g. nucleotide),
use a lot (e.g. &gt;10k) of small windows (e.g. `wsize=10`), as per default
settings. Otherwise the process can be lightened by using fewer bigger 
windows.</p></dd>


<dt>nwind</dt>
<dd><p>The number of random windows</p></dd>


<dt>peaks</dt>
<dd><p>A list of peaks (GRanges) for each experiment in `x`, or a
vector of paths to such files, or a single GRanges of peaks to use for 
MAnorm</p></dd>


<dt>trim</dt>
<dd><p>Amount of trimming when calculating means</p></dd>


<dt>method</dt>
<dd><p>Normalization method (see details)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A vector of normalization factors, or for the 'S3norm' and '2cLinear'
  methods, a numeric matrix with a number of rows equal to the length 
  of `x`, and two columns indicating the alpha and beta terms.</p>


<p>A vector of normalization factors or, for methods 'S3norm' and 
  '2cLinear', a matrix of per-sample normalization parameters.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The 'background' or 'SES' normalization method (they are synonyms here)
(Diaz et al., Stat Appl Gen et Mol Biol, 2012) assumes that the background
noise should on average be the same across experiments, an assumption that 
works well in practice when there are not very large differences in 
signal-to-noise ratio. The implementation uses the trimmed mean number of 
reads in random windows with non-zero counts.
The 'MAnorm' approach (Shao et al., Genome Biology 
2012) assumes that regions that are commonly enriched (i.e. common peaks) in 
two experiments should on average have the same signal in the two 
experiments. These methods then use linear scaling.
The 'S3norm' (Xiang et al., NAR 2020) and '2cLinear' methods try to 
normalize both simultaneously. S3norm does this in a log-linear fashion (as 
in the publication), while '2cLinear' does it on the original scale.</p>
    </div>
    <div class="section level2">
    <h2 id="functions">Functions<a class="anchor" aria-label="anchor" href="#functions"></a></h2>
    
<ul><li><p><code>bwNormFactors()</code>: deprecated in favor of getNormFactors</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># we get an example bigwig file, and use it twice:</span></span></span>
<span class="r-in"><span><span class="va">bw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example_atac.bw"</span>, package<span class="op">=</span><span class="st">"epiwraps"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>sample1<span class="op">=</span><span class="va">bw</span>, sample2<span class="op">=</span><span class="va">bw</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># we indicate to use only chr1, because it's the only one in the example file</span></span></span>
<span class="r-in"><span><span class="fu">getNormFactors</span><span class="op">(</span><span class="va">bw</span>, useSeqLevels<span class="op">=</span><span class="st">"1"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Comparing coverage in random regions...</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Some samples have less than 50 non-zero windows. Consider increasing the window size or (better) the number of windows.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> sample1 sample2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       1       1 </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre-Luc Germain, Emanuel Sonder.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

