<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ epiwraps</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">epiwraps</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.107</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/bam2bw.html">Generating bigwig tracks with bam2bw</a></li>
    <li><a class="dropdown-item" href="../articles/misc.html">Miscellaneous epiwraps functions</a></li>
    <li><a class="dropdown-item" href="../articles/multiRegionPlot.html">Visualizing signals across many regions</a></li>
    <li><a class="dropdown-item" href="../articles/normalization.html">Normalizing genomic signals</a></li>
    <li><a class="dropdown-item" href="../articles/singleRegionPlot.html">Visualizing signals in a single region</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="TSSenrichment.html">TSSenrichment()</a></code>

        </dt>
        <dd>TSSenrichment</dd>
      </dl><dl><dt>

          <code><a href="addAssayToESE.html">addAssayToESE()</a></code>

        </dt>
        <dd>addAssayToESE</dd>
      </dl><dl><dt>

          <code><a href="annotateRegions.html">annotateRegions()</a></code>

        </dt>
        <dd>annotateRegions</dd>
      </dl><dl><dt>

          <code><a href="bam2bw.html">bam2bw()</a></code>

        </dt>
        <dd>bam2bw: create a coverage bigwig file from an alignment bam file.</dd>
      </dl><dl><dt>

          <code><a href="bamChrChunkApply.html">bamChrChunkApply()</a></code>

        </dt>
        <dd>bamChrChunkApply</dd>
      </dl><dl><dt>

          <code><a href="breakStrings.html">breakStrings()</a></code>

        </dt>
        <dd>breakStrings</dd>
      </dl><dl><dt>

          <code><a href="clusterSignalMatrices.html">clusterSignalMatrices()</a></code>

        </dt>
        <dd>clusterSignalMatrices: clusters the regions of a (set of) signal matrices.</dd>
      </dl><dl><dt>

          <code><a href="computeDeviationsFaster.html">computeDeviationsFaster()</a></code>

        </dt>
        <dd>computeDeviationsFaster</dd>
      </dl><dl><dt>

          <code><a href="estimateFragSize.html">estimateFragSize()</a></code>

        </dt>
        <dd>estimateFragSize</dd>
      </dl><dl><dt>

          <code><a href="exampleDNAme.html">exampleDNAme</a></code> <code><a href="exampleDNAme.html">geneBodies</a></code>

        </dt>
        <dd>Example DNAme data on some active gene bodies</dd>
      </dl><dl><dt>

          <code><a href="exampleESE.html">exampleESE</a></code>

        </dt>
        <dd>Example EnrichmentSE object</dd>
      </dl><dl><dt>

          <code><a href="formatGenomicDist.html">formatGenomicDist()</a></code>

        </dt>
        <dd>formatGenomicDist</dd>
      </dl><dl><dt>

          <code><a href="frag2bw.html">frag2bw()</a></code>

        </dt>
        <dd>frag2bw</dd>
      </dl><dl><dt>

          <code><a href="fragSizesDist.html">fragSizesDist()</a></code>

        </dt>
        <dd>fragSizesDist</dd>
      </dl><dl><dt>

          <code><a href="getCounts.html">getCounts()</a></code>

        </dt>
        <dd>getCounts</dd>
      </dl><dl><dt>

          <code><a href="getCovStats.html">getCovStats()</a></code>

        </dt>
        <dd>getCovStats</dd>
      </dl><dl><dt>

          <code><a href="getNormFactors.html">getNormFactors()</a></code> <code><a href="getNormFactors.html">bwNormFactors()</a></code>

        </dt>
        <dd>getNormFactors : estimate normalization factors from genomic signal files</dd>
      </dl><dl><dt>

          <code><a href="getSignalMatrices.html">getSignalMatrices()</a></code>

        </dt>
        <dd>getSignalMatrices</dd>
      </dl><dl><dt>

          <code><a href="importBedlike.html">importBedlike()</a></code>

        </dt>
        <dd>importBedlike</dd>
      </dl><dl><dt>

          <code><a href="inject.html">inject()</a></code>

        </dt>
        <dd>Inject (insert) values at positions in a vector</dd>
      </dl><dl><dt>

          <code><a href="meltSignals.html">meltSignals()</a></code>

        </dt>
        <dd>meltSignals</dd>
      </dl><dl><dt>

          <code><a href="mergeSignalMatrices.html">mergeSignalMatrices()</a></code>

        </dt>
        <dd>mergeSignalMatrices: aggregates two or more signal matrices.</dd>
      </dl><dl><dt>

          <code><a href="ml2ESE.html">ml2ESE()</a></code>

        </dt>
        <dd>Creates an EnrichmentSE from a list of normalizedMatrix objects</dd>
      </dl><dl><dt>

          <code><a href="motifFootprint.html">motifFootprint()</a></code>

        </dt>
        <dd>motifFootprint</dd>
      </dl><dl><dt>

          <code><a href="plotCorFromCovStats.html">plotCorFromCovStats()</a></code>

        </dt>
        <dd>plotCorFromCovStats</dd>
      </dl><dl><dt>

          <code><a href="plotCovStats.html">plotCovStats()</a></code>

        </dt>
        <dd>plotCovStats</dd>
      </dl><dl><dt>

          <code><a href="plotEnrichedHeatmaps.html">plotEnrichedHeatmaps()</a></code>

        </dt>
        <dd>plotEnrichedHeatmaps: Plots heatmaps of signals around a set of regions</dd>
      </dl><dl><dt>

          <code><a href="plotSignalTracks.html">plotSignalTracks()</a></code>

        </dt>
        <dd>plotSignalTracks</dd>
      </dl><dl><dt>

          <code><a href="reduceWithResplit.html">reduceWithResplit()</a></code>

        </dt>
        <dd>Merge regions, re-splitting large merges using local overlap minima</dd>
      </dl><dl><dt>

          <code><a href="regionCAT.html">regionCAT()</a></code>

        </dt>
        <dd>regionCAT</dd>
      </dl><dl><dt>

          <code><a href="regionOverlaps.html">regionOverlaps()</a></code>

        </dt>
        <dd>regionOverlaps</dd>
      </dl><dl><dt>

          <code><a href="regionUpset.html">regionUpset()</a></code>

        </dt>
        <dd>regionUpset</dd>
      </dl><dl><dt>

          <code><a href="renormalizeSignalMatrices.html">renormalizeBorders()</a></code> <code><a href="renormalizeSignalMatrices.html">renormalizeSignalMatrices()</a></code>

        </dt>
        <dd>renormalizeSignalMatrices</dd>
      </dl><dl><dt>

          <code><a href="resizeMatrix.html">resizeMatrix()</a></code>

        </dt>
        <dd>resize a numeric matrix to given dimensions</dd>
      </dl><dl><dt>

          <code><a href="showTrackInfo.html">showTrackInfo()</a></code>

        </dt>
        <dd>showTrackInfo</dd>
      </dl><dl><dt>

          <code><a href="signal2Matrix.html">signal2Matrix()</a></code>

        </dt>
        <dd>signal2Matrix: reads the signals in/around a set of genomic regions.</dd>
      </dl><dl><dt>

          <code><a href="signalsAcrossSamples.html">signalsAcrossSamples()</a></code>

        </dt>
        <dd>signalsAcrossSamples</dd>
      </dl><dl><dt>

          <code><a href="tabixChrApply.html">tabixChrApply()</a></code>

        </dt>
        <dd>tabixChrApply</dd>
      </dl><dl><dt>

          <code><a href="tileRle.html">tileRle()</a></code>

        </dt>
        <dd>tileRle</dd>
      </dl><dl><dt>

          <code><a href="views2Matrix.html">views2Matrix()</a></code>

        </dt>
        <dd>views2Matrix</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre-Luc Germain, Emanuel Sonder.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

