<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>plotSignalTracks — plotSignalTracks • epiwraps</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="plotSignalTracks — plotSignalTracks"><meta name="description" content="A wrapper around `Gviz` for quick plotting of genomic signals in a single
region."><meta property="og:description" content="A wrapper around `Gviz` for quick plotting of genomic signals in a single
region."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">epiwraps</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.105</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/bam2bw.html">Generating bigwig tracks with bam2bw</a></li>
    <li><a class="dropdown-item" href="../articles/misc.html">Miscellaneous epiwraps functions</a></li>
    <li><a class="dropdown-item" href="../articles/multiRegionPlot.html">Visualizing signals across many regions</a></li>
    <li><a class="dropdown-item" href="../articles/normalization.html">Normalizing genomic signals</a></li>
    <li><a class="dropdown-item" href="../articles/singleRegionPlot.html">Visualizing signals in a single region</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>plotSignalTracks</h1>

      <div class="d-none name"><code>plotSignalTracks.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A wrapper around `Gviz` for quick plotting of genomic signals in a single
region.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotSignalTracks</span><span class="op">(</span></span>
<span>  files <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="va">region</span>,</span>
<span>  ensdb <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  colors <span class="op">=</span> <span class="st">"darkblue"</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"histogram"</span>,</span>
<span>  genomeAxis <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>  extend <span class="op">=</span> <span class="fl">0.15</span>,</span>
<span>  aggregation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mean"</span>, <span class="st">"median"</span>, <span class="st">"sum"</span>, <span class="st">"max"</span>, <span class="st">"min"</span>, <span class="st">"heatmap"</span>, <span class="st">"overlay"</span>,</span>
<span>    <span class="st">"heatmap+mean"</span><span class="op">)</span>,</span>
<span>  transcripts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"collapsed"</span>, <span class="st">"full"</span>, <span class="st">"coding"</span>, <span class="st">"none"</span><span class="op">)</span>,</span>
<span>  genes.params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>col.line <span class="op">=</span> <span class="st">"grey40"</span>, col <span class="op">=</span> <span class="cn">NULL</span>, fill <span class="op">=</span> <span class="st">"#000000"</span>, rotation.title <span class="op">=</span></span>
<span>    <span class="fl">0</span><span class="op">)</span>,</span>
<span>  align.params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>color <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span>,</span>
<span>  tracks.params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  extraTracks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  background.title <span class="op">=</span> <span class="st">"white"</span>,</span>
<span>  col.axis <span class="op">=</span> <span class="st">"grey40"</span>,</span>
<span>  bed.rotation.title <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  col.title <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  cex.title <span class="op">=</span> <span class="fl">0.65</span>,</span>
<span>  overlay.alpha <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  normFactors <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-files">files<a class="anchor" aria-label="anchor" href="#arg-files"></a></dt>
<dd><p>A named list or vector of paths to signal files  (e.g.
bigwig/bam, but also bed files). If a list, list elements will be overlaid
or aggregated (depending on the `aggregation` argument). Formats accepted by
<code><a href="https://rdrr.io/pkg/Gviz/man/DataTrack-class.html" class="external-link">DataTrack</a></code>'s `range` argument are also accepted. Can also
include `GRanges ` objects (which will be plotted as
<code><a href="https://rdrr.io/pkg/Gviz/man/AnnotationTrack-class.html" class="external-link">AnnotationTrack</a></code>) or objects inheriting the
<code>GdObject</code> class (i.e. any `Gviz` track object).</p></dd>


<dt id="arg-region">region<a class="anchor" aria-label="anchor" href="#arg-region"></a></dt>
<dd><p>A genomic region, either as a `GRanges` object or as a string
(i.e. `region="chr5:10000-12000`). Alternatively, if `ensdb` is provided, a
gene name can be given, and the gene's coordinates will be used as region.</p></dd>


<dt id="arg-ensdb">ensdb<a class="anchor" aria-label="anchor" href="#arg-ensdb"></a></dt>
<dd><p>An optional <code><a href="https://rdrr.io/pkg/ensembldb/man/EnsDb.html" class="external-link">EnsDb</a></code> object form which
to grab transcripts.</p></dd>


<dt id="arg-colors">colors<a class="anchor" aria-label="anchor" href="#arg-colors"></a></dt>
<dd><p>Signal color(s); will be recycled for elements of `files`</p></dd>


<dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>Signal plot type(s); will be recycled for elements of `files`.
This is ignored for bed-like files, which are shown as
<code><a href="https://rdrr.io/pkg/Gviz/man/AnnotationTrack-class.html" class="external-link">AnnotationTrack</a></code>. See the `type` options of
<code><a href="https://rdrr.io/pkg/Gviz/man/DataTrack-class.html" class="external-link">DataTrack</a></code>. In addition to these options, the type
'alignments' can be given for bam files, which will display them as
<code><a href="https://rdrr.io/pkg/Gviz/man/AlignmentsTrack-class.html" class="external-link">AlignmentsTrack</a></code>.</p></dd>


<dt id="arg-genomeaxis">genomeAxis<a class="anchor" aria-label="anchor" href="#arg-genomeaxis"></a></dt>
<dd><p>Whether to plot a genome axis. Alternatively, a numeric
scalar between 0 and 1 can be given, in which case a scale will be
plotted of this relative size.</p></dd>


<dt id="arg-extend">extend<a class="anchor" aria-label="anchor" href="#arg-extend"></a></dt>
<dd><p>Either an integer or vector of two integers indicating the
number of base pairs by which to extent on either side. If `extend`&lt;=1, this
will be interpreted as a fraction of the plotted region.</p></dd>


<dt id="arg-aggregation">aggregation<a class="anchor" aria-label="anchor" href="#arg-aggregation"></a></dt>
<dd><p>Method for aggregation data tracks, one of: 'mean'
(default), 'median', 'max', 'overlay', 'heatmap', or 'heatmap+mean'. The
latter will create a mean plot of type `type` followed by a heatmap.</p></dd>


<dt id="arg-transcripts">transcripts<a class="anchor" aria-label="anchor" href="#arg-transcripts"></a></dt>
<dd><p>Whether to show transcripts (reguires `ensdb`) as "full",
"collapsed" (default), "coding" (only coding transcripts) or "none".
Alternatively, can be a custom <code><a href="https://rdrr.io/pkg/Gviz/man/GeneRegionTrack-class.html" class="external-link">GeneRegionTrack</a></code> object.</p></dd>


<dt id="arg-genes-params">genes.params<a class="anchor" aria-label="anchor" href="#arg-genes-params"></a></dt>
<dd><p>Named list of parameters passed to
<code><a href="https://rdrr.io/pkg/Gviz/man/GeneRegionTrack-class.html" class="external-link">GeneRegionTrack</a></code>.</p></dd>


<dt id="arg-align-params">align.params<a class="anchor" aria-label="anchor" href="#arg-align-params"></a></dt>
<dd><p>Named list of parameters passed to
<code><a href="https://rdrr.io/pkg/Gviz/man/AlignmentsTrack-class.html" class="external-link">AlignmentsTrack</a></code>. Only used for plotting bam files with
`type="alignments"`.</p></dd>


<dt id="arg-tracks-params">tracks.params<a class="anchor" aria-label="anchor" href="#arg-tracks-params"></a></dt>
<dd><p>Named list of parameters passed to
<code><a href="https://rdrr.io/pkg/Gviz/man/DataTrack-class.html" class="external-link">DataTrack</a></code>.</p></dd>


<dt id="arg-extratracks">extraTracks<a class="anchor" aria-label="anchor" href="#arg-extratracks"></a></dt>
<dd><p>List of extra custom tracks to be plotted.</p></dd>


<dt id="arg-background-title">background.title<a class="anchor" aria-label="anchor" href="#arg-background-title"></a></dt>
<dd><p>The background color of the track titles.</p></dd>


<dt id="arg-col-axis">col.axis<a class="anchor" aria-label="anchor" href="#arg-col-axis"></a></dt>
<dd><p>The color of the axes.</p></dd>


<dt id="arg-bed-rotation-title">bed.rotation.title<a class="anchor" aria-label="anchor" href="#arg-bed-rotation-title"></a></dt>
<dd><p>Rotation for track titles of bed files.</p></dd>


<dt id="arg-col-title">col.title<a class="anchor" aria-label="anchor" href="#arg-col-title"></a></dt>
<dd><p>The color of the track titles.</p></dd>


<dt id="arg-cex-title">cex.title<a class="anchor" aria-label="anchor" href="#arg-cex-title"></a></dt>
<dd><p>Expension factor for the font size of the track titles.</p></dd>


<dt id="arg-overlay-alpha">overlay.alpha<a class="anchor" aria-label="anchor" href="#arg-overlay-alpha"></a></dt>
<dd><p>Transparency (0 to 250) when overlaying tracks.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Passed to <code><a href="https://rdrr.io/pkg/Gviz/man/plotTracks.html" class="external-link">plotTracks</a></code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list of GenomeGraph tracks to be plotted.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># fetch path to example bigwig file:</span></span></span>
<span class="r-in"><span><span class="op">(</span><span class="va">bw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example_rna.bw"</span>, package<span class="op">=</span><span class="st">"epiwraps"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "/home/runner/work/_temp/Library/epiwraps/extdata/example_rna.bw"</span>
<span class="r-in"><span><span class="fu">plotSignalTracks</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>track1<span class="op">=</span><span class="va">bw</span><span class="op">)</span>, region<span class="op">=</span><span class="st">"8:22165140-22212326"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotSignalTracks-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># if we had an EnsDb object loaded, we could just input a gene instead of </span></span></span>
<span class="r-in"><span><span class="co"># coordinates, and the transcript models would automatically show (not run):</span></span></span>
<span class="r-in"><span><span class="co"># plotSignalTracks(list(track1=bw), region="BMP1", ensdb=ensdb)</span></span></span>
<span class="r-in"><span><span class="co"># show all transcript variants:</span></span></span>
<span class="r-in"><span><span class="co"># plotSignalTracks(list(tracks=bw), region="BMP1", ensdb=ensdb,</span></span></span>
<span class="r-in"><span><span class="co">#                  transcripts="full")</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre-Luc Germain, Emanuel Sonder.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

