<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>signal2Matrix: reads the signals in/around a set of genomic regions. — signal2Matrix • epiwraps</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="signal2Matrix: reads the signals in/around a set of genomic regions. — signal2Matrix"><meta name="description" content="Reads the signals from bam/bigwig files within and/or around (the centers of)
a set of regions, and outputs an EnrichmentSE object."><meta property="og:description" content="Reads the signals from bam/bigwig files within and/or around (the centers of)
a set of regions, and outputs an EnrichmentSE object."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">epiwraps</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.100</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/bam2bw.html">Generating bigwig tracks with bam2bw</a></li>
    <li><a class="dropdown-item" href="../articles/misc.html">Miscellaneous epiwraps functions</a></li>
    <li><a class="dropdown-item" href="../articles/multiRegionPlot.html">Visualizing signals across many regions</a></li>
    <li><a class="dropdown-item" href="../articles/normalization.html">Normalizing genomic signals</a></li>
    <li><a class="dropdown-item" href="../articles/singleRegionPlot.html">Visualizing signals in a single region</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>signal2Matrix: reads the signals in/around a set of genomic regions.</h1>

      <div class="d-none name"><code>signal2Matrix.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Reads the signals from bam/bigwig files within and/or around (the centers of)
a set of regions, and outputs an EnrichmentSE object.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">signal2Matrix</span><span class="op">(</span></span>
<span>  <span class="va">filepaths</span>,</span>
<span>  <span class="va">regions</span>,</span>
<span>  extend <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  w <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scaling <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scaledBins <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">extend</span><span class="op">)</span><span class="op">/</span><span class="va">w</span><span class="op">)</span>,</span>
<span>  type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"center"</span>, <span class="st">"scaled"</span><span class="op">)</span>,</span>
<span>  binMethod <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mean"</span>, <span class="st">"max"</span>, <span class="st">"min"</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  ret <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EnrichmentSE"</span>, <span class="st">"list"</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-filepaths">filepaths<a class="anchor" aria-label="anchor" href="#arg-filepaths"></a></dt>
<dd><p>A named vector of filepaths (e.g. to bigwig files; bam files
are also supported, but with limited functionalities). Can also be a named
list including a combination of paths to such files and `GRanges` object.
For `GRanges` objects, the `score` column will be used (absolute coverage
mode).</p></dd>


<dt id="arg-regions">regions<a class="anchor" aria-label="anchor" href="#arg-regions"></a></dt>
<dd><p>A `GRanges` of the regions/positions around which to plot, or
the path to a bed file of such regions. If `type="scaled"`, `regions` can
also be a `GRangesList`, in which case the coverage of the subregions will
be stiched together (e.g. for plotting exonic signal over transcripts).</p></dd>


<dt id="arg-extend">extend<a class="anchor" aria-label="anchor" href="#arg-extend"></a></dt>
<dd><p>Number of basepair to extend on either side of the regions.
Must be a multiple of `w`. Can also be an integer of length 2, indicating
the extension upstream and downstream.</p></dd>


<dt id="arg-w">w<a class="anchor" aria-label="anchor" href="#arg-w"></a></dt>
<dd><p>Bin width in number of nucleotides. Defaults to a width producing
200 bins over the whole extended range.</p></dd>


<dt id="arg-scaledbins">scaledBins<a class="anchor" aria-label="anchor" href="#arg-scaledbins"></a></dt>
<dd><p>The number of bins for the scale region (ignored if
`type="center"`)</p></dd>


<dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>Either 'center' (plots fixed-size region around the centers of
`regions`) or 'scaled' (scales the signal in `regions` and plot
surroundings)</p></dd>


<dt id="arg-binmethod">binMethod<a class="anchor" aria-label="anchor" href="#arg-binmethod"></a></dt>
<dd><p>Whether to compute the 'max' (default), 'mean' or 'min' per
bin.</p></dd>


<dt id="arg-bpparam">BPPARAM<a class="anchor" aria-label="anchor" href="#arg-bpparam"></a></dt>
<dd><p>A <code><a href="https://rdrr.io/pkg/BiocParallel/man/BiocParallelParam-class.html" class="external-link">BiocParallelParam</a></code> object, or
the number of threads to use to read and prepare the data. Note that the
rate-limiting process is reading from disk, so unless you have an unusually
fast disk, using multi-threading is actually likely to slow down rather
than speed up the process.</p></dd>


<dt id="arg-ret">ret<a class="anchor" aria-label="anchor" href="#arg-ret"></a></dt>
<dd><p>The type of output to return, either an "EnrichmentSE" object
(default), or a simple list of signal matrices ("list").</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Logical; whether to print processing information</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Passed to <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html" class="external-link">normalizeToMatrix</a></code> or
<code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/as.normalizedMatrix.html" class="external-link">as.normalizedMatrix</a></code>, or to
<code><a href="bam2bw.html">bam2bw</a></code> when reading bam files. For example, this can be used
to pass arguments to `normalizeToMatrix` such as `smooth=TRUE`.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list of `normalizeToMatrix` objects</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># we fetch the path to the example bigwig file:</span></span></span>
<span class="r-in"><span><span class="op">(</span><span class="va">bw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example_atac.bw"</span>, package<span class="op">=</span><span class="st">"epiwraps"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "/home/runner/work/_temp/Library/epiwraps/extdata/example_atac.bw"</span>
<span class="r-in"><span><span class="co"># we load example regions:</span></span></span>
<span class="r-in"><span><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">import</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example_peaks.bed"</span>, </span></span>
<span class="r-in"><span>                                           package<span class="op">=</span><span class="st">"epiwraps"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">regions</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 150</span>
<span class="r-in"><span><span class="co"># we obtain the matrix of the signal around the regions:</span></span></span>
<span class="r-in"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu">signal2Matrix</span><span class="op">(</span><span class="va">bw</span>, <span class="va">regions</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading /home/runner/work/_temp/Library/epiwraps/extdata/example_atac.bw</span>
<span class="r-in"><span><span class="co"># we can plot it with:</span></span></span>
<span class="r-in"><span><span class="fu"><a href="plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="signal2Matrix-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># we could also take a broader range around the center of the regions, and </span></span></span>
<span class="r-in"><span><span class="co"># use bigger bins:</span></span></span>
<span class="r-in"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu">signal2Matrix</span><span class="op">(</span><span class="va">bw</span>, <span class="va">regions</span>, extend<span class="op">=</span><span class="fl">2000</span>, w<span class="op">=</span><span class="fl">20</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading /home/runner/work/_temp/Library/epiwraps/extdata/example_atac.bw</span>
<span class="r-in"><span><span class="fu"><a href="plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="signal2Matrix-2.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre-Luc Germain, Emanuel Sonder.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

