<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="epiwraps">
<title>Generating bigwig tracks with bam2bw • epiwraps</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Generating bigwig tracks with bam2bw">
<meta property="og:description" content="epiwraps">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">epiwraps</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.83</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/bam2bw.html">Generating bigwig tracks with bam2bw</a>
    <a class="dropdown-item" href="../articles/misc.html">Miscellaneous epiwraps functions</a>
    <a class="dropdown-item" href="../articles/multiRegionPlot.html">Visualizing signals across many regions</a>
    <a class="dropdown-item" href="../articles/singleRegionPlot.html">Visualizing signals in a single region</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Generating bigwig tracks with bam2bw</h1>
                        <h4 data-toc-skip class="author">Pierre-Luc
Germain</h4>
            <address class="author_afil">
      D-HEST Institute for Neuroscience, ETHLab of Statistical
Bioinformatics, UZH<br><div class="d-none name"><code>bam2bw.Rmd</code></div>
    </address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>This vignette documents the use of the ‘bam2bw’ function of the
      epiwraps package, which generates bigwig tracks from alignment
      files in an efficient and flexible fashion.</p>
    </div>
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <code>bam2bw</code> function can be used to compute
per-nucleotide or per-bin coverage from alignments and save it to a
bigwig file. In this process, information about individual reads is
lost, but the produced signals are considerably more lightweight and
amenable to visualization. The bigwig format is readily queried from R
or compatible with a variety of tools, including genome browsers.</p>
<div class="section level3">
<h3 id="many-way-of-compiling-coverage">Many way of compiling coverage<a class="anchor" aria-label="anchor" href="#many-way-of-compiling-coverage"></a>
</h3>
<p>To introduce the different variations on coverage, let’s assume
you’ve go the following single-end reads:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ethz-ins.github.io/epiwraps/">epiwraps</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># we create some arbitrary genomic ranges</span></span>
<span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">30</span>,<span class="fl">70</span>,<span class="fl">120</span><span class="op">)</span>, width<span class="op">=</span><span class="fl">50</span><span class="op">)</span>, strand<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>,<span class="st">"+"</span>,<span class="st">"-"</span><span class="op">)</span>,</span>
<span>              seqlengths<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>chr1<span class="op">=</span><span class="fl">500</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotSignalTracks.html">plotSignalTracks</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>reads<span class="op">=</span><span class="va">gr</span><span class="op">)</span>, region<span class="op">=</span><span class="st">"chr1:1:180"</span>, extend<span class="op">=</span><span class="fl">0</span>, genomeAxis<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="bam2bw_files/figure-html/unnamed-chunk-2-1.png" width="384"></p>
<p>For testing purposes, we’ll save this as a bam file and index it:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".bam"</span><span class="op">)</span> <span class="co"># temp file name</span></span>
<span><span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">export</span><span class="op">(</span><span class="va">gr</span>, <span class="va">bam</span>, format<span class="op">=</span><span class="st">"bam"</span><span class="op">)</span></span>
<span><span class="fu">Rsamtools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/scanBam.html" class="external-link">indexBam</a></span><span class="op">(</span><span class="va">bam</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       /tmp/RtmpvbXrLt/file1ad071ec30f1.bam </span></span>
<span><span class="co">## "/tmp/RtmpvbXrLt/file1ad071ec30f1.bam.bai"</span></span></code></pre>
<p>Using these example reads, we can illustrate different ways of
computing coverages.</p>
<p>First, we can save coverage at different resolutions, from full
resolution (each nucleotide is a single bin) to larger bin sizes, the
latter giving smaller filesizes:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Full coverage with bin width of 1 nucleotide (i.e. full resolution)</span></span>
<span><span class="va">cov_full_bw1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".bw"</span><span class="op">)</span> <span class="co"># temp file name</span></span>
<span><span class="fu"><a href="../reference/bam2bw.html">bam2bw</a></span><span class="op">(</span><span class="va">bam</span>, <span class="va">cov_full_bw1</span>, binWidth<span class="op">=</span><span class="fl">1L</span>, scaling<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## `paired` not specified, assuming single-end reads. Set to paired='auto' to automatically detect.</span></span></code></pre>
<pre><code><span><span class="co">## Reading in signal...</span></span></code></pre>
<pre><code><span><span class="co">## Writing bigwig...</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Full coverage with larger bins</span></span>
<span><span class="va">cov_full_bw25</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".bw"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/bam2bw.html">bam2bw</a></span><span class="op">(</span><span class="va">bam</span>, <span class="va">cov_full_bw25</span>, binWidth<span class="op">=</span><span class="fl">25L</span>, scaling<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## `paired` not specified, assuming single-end reads. Set to paired='auto' to automatically detect.</span></span></code></pre>
<pre><code><span><span class="co">## Reading in signal...</span></span></code></pre>
<pre><code><span><span class="co">## Writing bigwig...</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSignalTracks.html">plotSignalTracks</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>reads<span class="op">=</span><span class="va">gr</span>, <span class="st">"binWidth=1"</span><span class="op">=</span><span class="va">cov_full_bw1</span>, <span class="st">"binWidth=25"</span><span class="op">=</span><span class="va">cov_full_bw25</span><span class="op">)</span>,</span>
<span>                 region<span class="op">=</span><span class="st">"chr1:1:180"</span>, extend<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="bam2bw_files/figure-html/unnamed-chunk-4-1.png" width="432"></p>
<p>Both tracks compile the number of reads that overlap each position,
but in the bottom track the signal is by chunks of 25 nucleotides. By
default, the maximum signal inside a bin is used, however it is possible
to change this:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using mean per bin:</span></span>
<span><span class="va">cov_full_bw25mean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".bw"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/bam2bw.html">bam2bw</a></span><span class="op">(</span><span class="va">bam</span>, <span class="va">cov_full_bw25mean</span>, binWidth<span class="op">=</span><span class="fl">25L</span>, binSummarization <span class="op">=</span> <span class="st">"mean"</span>, scaling<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## `paired` not specified, assuming single-end reads. Set to paired='auto' to automatically detect.</span></span></code></pre>
<pre><code><span><span class="co">## Reading in signal...</span></span></code></pre>
<pre><code><span><span class="co">## Writing bigwig...</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSignalTracks.html">plotSignalTracks</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>reads<span class="op">=</span><span class="va">gr</span>, <span class="st">"binWidth=1"</span><span class="op">=</span><span class="va">cov_full_bw1</span>, </span>
<span>                      <span class="st">"binWidth=25\n(max)"</span><span class="op">=</span><span class="va">cov_full_bw25</span>,</span>
<span>                      <span class="st">"binWidth=25\n(mean)"</span><span class="op">=</span><span class="va">cov_full_bw25mean</span><span class="op">)</span>,</span>
<span>                 region<span class="op">=</span><span class="st">"chr1:1:180"</span>, extend<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="bam2bw_files/figure-html/unnamed-chunk-5-1.png" width="432"></p>
<p>In most cases, single-end reads are just the beginning of the DNA
fragments obtained, and in most cases we know the average size of the
fragments (if not, see the <code>estimateFragSize</code> function, or
the simpler <code>estimate.mean.fraglen</code> function of the <em><a href="https://bioconductor.org/packages/3.18/chipseq" class="external-link">chipseq</a></em>
package). It is therefore common to extend reads to this size when
computing coverage, so as to obtain the number of fragments (rather than
reads) coverage each position. This can be done as follows:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Here the reads are 50bp, and we want to extend them to 100bp, hence _by_ 50:</span></span>
<span><span class="va">cov_full_ext</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".bw"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/bam2bw.html">bam2bw</a></span><span class="op">(</span><span class="va">bam</span>, <span class="va">cov_full_ext</span>, binWidth<span class="op">=</span><span class="fl">1L</span>, extend<span class="op">=</span><span class="fl">50L</span>, scaling<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## `paired` not specified, assuming single-end reads. Set to paired='auto' to automatically detect.</span></span></code></pre>
<pre><code><span><span class="co">## Reading in signal...</span></span></code></pre>
<pre><code><span><span class="co">## Writing bigwig...</span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSignalTracks.html">plotSignalTracks</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>reads<span class="op">=</span><span class="va">gr</span>, <span class="st">"no extension"</span><span class="op">=</span><span class="va">cov_full_bw1</span>, </span>
<span>                      <span class="st">"read extension"</span><span class="op">=</span><span class="va">cov_full_ext</span><span class="op">)</span>,</span>
<span>                 region<span class="op">=</span><span class="st">"chr1:1:190"</span>, extend<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="bam2bw_files/figure-html/unnamed-chunk-6-1.png" width="432"></p>
<p>Instead of computing coverage, we could compute the number of reads
starting, ending, or being centered at each position:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Here the reads are 50bp, and we want to extend them to 100bp, hence _by_ 50:</span></span>
<span><span class="va">cov_start</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".bw"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/bam2bw.html">bam2bw</a></span><span class="op">(</span><span class="va">bam</span>, <span class="va">cov_start</span>, binWidth<span class="op">=</span><span class="fl">1L</span>, extend<span class="op">=</span><span class="fl">50L</span>, scaling<span class="op">=</span><span class="cn">FALSE</span>, type<span class="op">=</span><span class="st">"start"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## `paired` not specified, assuming single-end reads. Set to paired='auto' to automatically detect.</span></span></code></pre>
<pre><code><span><span class="co">## Reading in signal...</span></span></code></pre>
<pre><code><span><span class="co">## Writing bigwig...</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cov_center</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".bw"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/bam2bw.html">bam2bw</a></span><span class="op">(</span><span class="va">bam</span>, <span class="va">cov_center</span>, binWidth<span class="op">=</span><span class="fl">1L</span>, extend<span class="op">=</span><span class="fl">50L</span>, scaling<span class="op">=</span><span class="cn">FALSE</span>, type<span class="op">=</span><span class="st">"center"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## `paired` not specified, assuming single-end reads. Set to paired='auto' to automatically detect.</span></span></code></pre>
<pre><code><span><span class="co">## Reading in signal...</span></span></code></pre>
<pre><code><span><span class="co">## Writing bigwig...</span></span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSignalTracks.html">plotSignalTracks</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>reads<span class="op">=</span><span class="va">gr</span>, <span class="st">"type=full"</span><span class="op">=</span><span class="va">cov_full_bw1</span>, </span>
<span>                      <span class="st">"type=start"</span><span class="op">=</span><span class="va">cov_start</span>, <span class="st">"type=center"</span><span class="op">=</span><span class="va">cov_center</span><span class="op">)</span>,</span>
<span>                 region<span class="op">=</span><span class="st">"chr1:1:190"</span>, extend<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="bam2bw_files/figure-html/unnamed-chunk-7-1.png" width="432">
Note that when extending reads, as in this case, the position
(e.g. “center”) are relative to the extended read (i.e. extension is
applied first).</p>
</div>
<div class="section level3">
<h3 id="example-heatmaps-created-using-different-bigwig-generation-procedures">Example heatmaps created using different bigwig generation
procedures<a class="anchor" aria-label="anchor" href="#example-heatmaps-created-using-different-bigwig-generation-procedures"></a>
</h3>
<p>The following figure, created using <code>epiwraps</code> (see
<a href="multiRegionPlot.html">the vignette on generating such
plots</a>), represent chromatin accessibility (ATAC-seq) signals around
bound CTCF motifs in T-cells. The different signals are based on
different bigwig files derived from the same bam file using the
functions described above.</p>
<p><img src="example_bigwigs_hm.png" width="700"></p>
<ul>
<li>The first heatmap (‘full coverage’) was generated with default
parameter, and is the fragment coverage (i.e. how many fragments overlap
any given location).</li>
<li>The second heatmap shows the fragment of sizes compatible with
mono-nucleosomes, resizing fragments from their centers. The exact
arguments used were
<code>shift=c(4L,-5L), minFragLength=147, maxFragLength=230, type="center", extend=10L</code>.
Using this we can see nucleosomes well-positioned at some distance from
CTCF binding sites, but not nucleosome at the bound site itself.</li>
<li>The third shows the coverage of nucleosome-free fragments, in this
case it was used with
<code>shift=c(4L,-5L), maxFragLength=120</code>.</li>
<li>The fourth shows where the transposase inserted itself, and was
generated with
<code>shift=c(4L,-5L), binWidth=1L, maxFragLength=120, type="ends"</code>.
Using this we can see a nice footprint protected from the transposase by
CTCF binding.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="working-with-fragment-files-as-an-input">Working with fragment files as an input<a class="anchor" aria-label="anchor" href="#working-with-fragment-files-as-an-input"></a>
</h2>
<p>If you use fragment files (preferably tabix-indexed) rather than bam
files as input, you can still perform most of the above tasks. See the
<code><a href="../reference/frag2bw.html">?frag2bw</a></code> function for more information.</p>
<p><br><br></p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.3 (2024-02-29)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] epiwraps_0.99.83            EnrichedHeatmap_1.32.0     </span></span>
<span><span class="co">##  [3] ComplexHeatmap_2.18.0       SummarizedExperiment_1.32.0</span></span>
<span><span class="co">##  [5] Biobase_2.62.0              GenomicRanges_1.54.1       </span></span>
<span><span class="co">##  [7] GenomeInfoDb_1.38.7         IRanges_2.36.0             </span></span>
<span><span class="co">##  [9] S4Vectors_0.40.2            BiocGenerics_0.48.1        </span></span>
<span><span class="co">## [11] MatrixGenerics_1.14.0       matrixStats_1.2.0          </span></span>
<span><span class="co">## [13] BiocStyle_2.30.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3       rstudioapi_0.15.0        jsonlite_1.8.8          </span></span>
<span><span class="co">##   [4] shape_1.4.6.1            magrittr_2.0.3           GenomicFeatures_1.54.3  </span></span>
<span><span class="co">##   [7] rmarkdown_2.26           GlobalOptions_0.1.2      fs_1.6.3                </span></span>
<span><span class="co">##  [10] BiocIO_1.12.0            zlibbioc_1.48.0          ragg_1.3.0              </span></span>
<span><span class="co">##  [13] vctrs_0.6.5              memoise_2.0.1            Rsamtools_2.18.0        </span></span>
<span><span class="co">##  [16] RCurl_1.98-1.14          base64enc_0.1-3          htmltools_0.5.7         </span></span>
<span><span class="co">##  [19] S4Arrays_1.2.1           progress_1.2.3           curl_5.2.1              </span></span>
<span><span class="co">##  [22] SparseArray_1.2.4        Formula_1.2-5            sass_0.4.8              </span></span>
<span><span class="co">##  [25] bslib_0.6.1              htmlwidgets_1.6.4        desc_1.4.3              </span></span>
<span><span class="co">##  [28] plyr_1.8.9               Gviz_1.46.1              cachem_1.0.8            </span></span>
<span><span class="co">##  [31] GenomicAlignments_1.38.2 lifecycle_1.0.4          iterators_1.0.14        </span></span>
<span><span class="co">##  [34] pkgconfig_2.0.3          Matrix_1.6-5             R6_2.5.1                </span></span>
<span><span class="co">##  [37] fastmap_1.1.1            GenomeInfoDbData_1.2.11  clue_0.3-65             </span></span>
<span><span class="co">##  [40] digest_0.6.35            colorspace_2.1-0         AnnotationDbi_1.64.1    </span></span>
<span><span class="co">##  [43] textshaping_0.3.7        Hmisc_5.1-2              RSQLite_2.3.5           </span></span>
<span><span class="co">##  [46] filelock_1.0.3           fansi_1.0.6              httr_1.4.7              </span></span>
<span><span class="co">##  [49] abind_1.4-5              compiler_4.3.3           bit64_4.0.5             </span></span>
<span><span class="co">##  [52] doParallel_1.0.17        backports_1.4.1          htmlTable_2.4.2         </span></span>
<span><span class="co">##  [55] BiocParallel_1.36.0      DBI_1.2.2                UpSetR_1.4.0            </span></span>
<span><span class="co">##  [58] highr_0.10               biomaRt_2.58.2           rappdirs_0.3.3          </span></span>
<span><span class="co">##  [61] DelayedArray_0.28.0      rjson_0.2.21             tools_4.3.3             </span></span>
<span><span class="co">##  [64] foreign_0.8-86           nnet_7.3-19              glue_1.7.0              </span></span>
<span><span class="co">##  [67] restfulr_0.0.15          checkmate_2.3.1          cluster_2.1.6           </span></span>
<span><span class="co">##  [70] generics_0.1.3           gtable_0.3.4             BSgenome_1.70.2         </span></span>
<span><span class="co">##  [73] ensembldb_2.26.0         data.table_1.15.2        hms_1.1.3               </span></span>
<span><span class="co">##  [76] xml2_1.3.6               utf8_1.2.4               XVector_0.42.0          </span></span>
<span><span class="co">##  [79] foreach_1.5.2            pillar_1.9.0             stringr_1.5.1           </span></span>
<span><span class="co">##  [82] circlize_0.4.16          dplyr_1.1.4              BiocFileCache_2.10.1    </span></span>
<span><span class="co">##  [85] lattice_0.22-5           deldir_2.0-4             rtracklayer_1.62.0      </span></span>
<span><span class="co">##  [88] bit_4.0.5                biovizBase_1.50.0        tidyselect_1.2.1        </span></span>
<span><span class="co">##  [91] locfit_1.5-9.9           pbapply_1.7-2            Biostrings_2.70.2       </span></span>
<span><span class="co">##  [94] knitr_1.45               gridExtra_2.3            bookdown_0.38           </span></span>
<span><span class="co">##  [97] ProtGenerics_1.34.0      xfun_0.42                stringi_1.8.3           </span></span>
<span><span class="co">## [100] lazyeval_0.2.2           yaml_2.3.8               evaluate_0.23           </span></span>
<span><span class="co">## [103] codetools_0.2-19         interp_1.1-6             GenomicFiles_1.38.0     </span></span>
<span><span class="co">## [106] tibble_3.2.1             BiocManager_1.30.22      cli_3.6.2               </span></span>
<span><span class="co">## [109] rpart_4.1.23             systemfonts_1.0.6        munsell_0.5.0           </span></span>
<span><span class="co">## [112] jquerylib_0.1.4          dichromat_2.0-0.1        Rcpp_1.0.12             </span></span>
<span><span class="co">## [115] dbplyr_2.4.0             png_0.1-8                XML_3.99-0.16.1         </span></span>
<span><span class="co">## [118] parallel_4.3.3           pkgdown_2.0.7            ggplot2_3.5.0           </span></span>
<span><span class="co">## [121] blob_1.2.4               prettyunits_1.2.0        jpeg_0.1-10             </span></span>
<span><span class="co">## [124] latticeExtra_0.6-30      AnnotationFilter_1.26.0  bitops_1.0-7            </span></span>
<span><span class="co">## [127] viridisLite_0.4.2        VariantAnnotation_1.48.1 scales_1.3.0            </span></span>
<span><span class="co">## [130] purrr_1.0.2              crayon_1.5.2             GetoptLong_1.0.5        </span></span>
<span><span class="co">## [133] rlang_1.1.3              cowplot_1.1.3            KEGGREST_1.42.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Pierre-Luc Germain, Emanuel Sonder.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
