<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Miscellaneous epiwraps functions • epiwraps</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Miscellaneous epiwraps functions">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">epiwraps</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.106</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/bam2bw.html">Generating bigwig tracks with bam2bw</a></li>
    <li><a class="dropdown-item" href="../articles/misc.html">Miscellaneous epiwraps functions</a></li>
    <li><a class="dropdown-item" href="../articles/multiRegionPlot.html">Visualizing signals across many regions</a></li>
    <li><a class="dropdown-item" href="../articles/normalization.html">Normalizing genomic signals</a></li>
    <li><a class="dropdown-item" href="../articles/singleRegionPlot.html">Visualizing signals in a single region</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Miscellaneous epiwraps functions</h1>
                        <h4 data-toc-skip class="author">Pierre-Luc
Germain</h4>
            <address class="author_afil">
      Lab of Statistical Bioinformatics, University of Zürich;D-HEST
Institute for Neuroscience, ETH Zürich,
Switzerland<br><div class="d-none name"><code>misc.Rmd</code></div>
    </address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>This vignette introduces some more isolated epiwraps
      functions.</p>
    </div>
    
<div class="section level2">
<h2 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h2>
<div class="section level3">
<h3 id="coverage-statistics">Coverage statistics<a class="anchor" aria-label="anchor" href="#coverage-statistics"></a>
</h3>
<p>Coverage statistics give an overview of how the reads are distributed
across the genome (or more precisely, across a large number of random
regions). The <code>getCovStats</code> will compute such statistics from
bam or bigwig files (from bigwig files will be considerably faster, but
if the files are normalized the coverage/density will be relative).</p>
<p>Because our example data spans only part of a chromosome, we’ll
exclude completely empty regions using the <code>exclude</code>
parameter, which would normally be used to exclude regions likely to be
technical artefacts (e.g. blacklisted regions).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ethz-ins.github.io/epiwraps/">epiwraps</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># get the path to an example bigwig file:</span></span>
<span><span class="va">bwf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example_atac.bw"</span>, package<span class="op">=</span><span class="st">"epiwraps"</span><span class="op">)</span></span>
<span><span class="va">cs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCovStats.html">getCovStats</a></span><span class="op">(</span><span class="va">bwf</span>, exclude<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"1"</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4300000</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotCovStats.html">plotCovStats</a></span><span class="op">(</span><span class="va">cs</span><span class="op">)</span></span></code></pre></div>
<p><img src="misc_files/figure-html/covstats-1.png" width="700"></p>
<p>Panel A shows the proportion of sampled regions which are above a
certain read density (relative because this is a normalized bigwig file,
would be coverage otherwise). This shows us, for example, that as
expected only a minority of regions have any reads at all (indicating
that the reads are not randomly distributed). Panel B is what is
sometimes called a fingerprint plot. It similarly shows us that the
reads are concentrated in very few regions, since the vast majority of
regions have only a very low fraction of the coverage of a few
high-density regions. Randomly distributed reads would go along the
diagonal, but one normally has a curve somewhere between this line and
the lower-right corner – the farther away from the diagonal, to more
strongly enriched the data is.</p>
<p>This can be done for multiple files simultaneously. If we have
several files, we can also use the coverage in the random windows to
computer their similarity (see <code><a href="../reference/plotCorFromCovStats.html">?plotCorFromCovStats</a></code>).</p>
</div>
<div class="section level3">
<h3 id="fragment-length-distributions">Fragment length distributions<a class="anchor" aria-label="anchor" href="#fragment-length-distributions"></a>
</h3>
<p>Given one or more paired-end bam files, we can extract and plot the
fragment length distribution using:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fragSizesDist.html">fragSizesDist</a></span><span class="op">(</span><span class="va">bam</span>, what<span class="op">=</span><span class="fl">100</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="tss-enrichment">TSS enrichment<a class="anchor" aria-label="anchor" href="#tss-enrichment"></a>
</h3>
<p>The TSS enrichment can also be calculated and plotted using the
<code>TSSenrichment</code> function.</p>
</div>
</div>
<div class="section level2">
<h2 id="peak-calling">Peak calling<a class="anchor" aria-label="anchor" href="#peak-calling"></a>
</h2>
<p>A very experimental peak calling function can be used, either against
an input control or against local or global backgrounds:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/callPeaks.html">callPeaks</a></span><span class="op">(</span><span class="va">bam</span>, fragLength<span class="op">=</span><span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<p>Note that this function is still under heavy development, and its
usage at the moment is discouraged!</p>
</div>
<div class="section level2">
<h2 id="region-overlapping">Region overlapping<a class="anchor" aria-label="anchor" href="#region-overlapping"></a>
</h2>
<p>The <code>GenomicRanges</code> package offers fast and powerful
functions for overlapping genomic regions. <code>epiwraps</code>
includes wrappers around those for common tasks, such as calculating and
visualizing overlaps across multiple sets of regions (see
<code><a href="../reference/regionUpset.html">?regionUpset</a></code>, <code><a href="../reference/regionOverlaps.html">?regionOverlaps</a></code>, and
<code><a href="../reference/regionCAT.html">?regionCAT</a></code>).</p>
<p><br><br></p>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.1 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] epiwraps_0.99.106           EnrichedHeatmap_1.36.0     </span></span>
<span><span class="co">##  [3] ComplexHeatmap_2.22.0       SummarizedExperiment_1.36.0</span></span>
<span><span class="co">##  [5] Biobase_2.66.0              GenomicRanges_1.58.0       </span></span>
<span><span class="co">##  [7] GenomeInfoDb_1.42.3         IRanges_2.40.1             </span></span>
<span><span class="co">##  [9] S4Vectors_0.44.0            BiocGenerics_0.52.0        </span></span>
<span><span class="co">## [11] MatrixGenerics_1.18.1       matrixStats_1.5.0          </span></span>
<span><span class="co">## [13] BiocStyle_2.34.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3       rstudioapi_0.17.1        jsonlite_1.9.0          </span></span>
<span><span class="co">##   [4] shape_1.4.6.1            magrittr_2.0.3           GenomicFeatures_1.58.0  </span></span>
<span><span class="co">##   [7] farver_2.1.2             rmarkdown_2.29           GlobalOptions_0.1.2     </span></span>
<span><span class="co">##  [10] fs_1.6.5                 BiocIO_1.16.0            zlibbioc_1.52.0         </span></span>
<span><span class="co">##  [13] ragg_1.3.3               vctrs_0.6.5              memoise_2.0.1           </span></span>
<span><span class="co">##  [16] Rsamtools_2.22.0         RCurl_1.98-1.16          base64enc_0.1-3         </span></span>
<span><span class="co">##  [19] htmltools_0.5.8.1        S4Arrays_1.6.0           progress_1.2.3          </span></span>
<span><span class="co">##  [22] curl_6.2.1               SparseArray_1.6.1        Formula_1.2-5           </span></span>
<span><span class="co">##  [25] sass_0.4.9               bslib_0.9.0              htmlwidgets_1.6.4       </span></span>
<span><span class="co">##  [28] desc_1.4.3               plyr_1.8.9               Gviz_1.50.0             </span></span>
<span><span class="co">##  [31] httr2_1.1.0              cachem_1.1.0             GenomicAlignments_1.42.0</span></span>
<span><span class="co">##  [34] lifecycle_1.0.4          iterators_1.0.14         pkgconfig_2.0.3         </span></span>
<span><span class="co">##  [37] Matrix_1.7-1             R6_2.6.1                 fastmap_1.2.0           </span></span>
<span><span class="co">##  [40] GenomeInfoDbData_1.2.13  clue_0.3-66              digest_0.6.37           </span></span>
<span><span class="co">##  [43] colorspace_2.1-1         AnnotationDbi_1.68.0     textshaping_1.0.0       </span></span>
<span><span class="co">##  [46] Hmisc_5.2-2              RSQLite_2.3.9            labeling_0.4.3          </span></span>
<span><span class="co">##  [49] filelock_1.0.3           httr_1.4.7               abind_1.4-8             </span></span>
<span><span class="co">##  [52] compiler_4.4.2           withr_3.0.2              bit64_4.6.0-1           </span></span>
<span><span class="co">##  [55] doParallel_1.0.17        backports_1.5.0          htmlTable_2.4.3         </span></span>
<span><span class="co">##  [58] BiocParallel_1.40.0      DBI_1.2.3                UpSetR_1.4.0            </span></span>
<span><span class="co">##  [61] biomaRt_2.62.1           rappdirs_0.3.3           DelayedArray_0.32.0     </span></span>
<span><span class="co">##  [64] rjson_0.2.23             tools_4.4.2              foreign_0.8-87          </span></span>
<span><span class="co">##  [67] nnet_7.3-19              glue_1.8.0               restfulr_0.0.15         </span></span>
<span><span class="co">##  [70] checkmate_2.3.2          cluster_2.1.6            generics_0.1.3          </span></span>
<span><span class="co">##  [73] gtable_0.3.6             BSgenome_1.74.0          ensembldb_2.30.0        </span></span>
<span><span class="co">##  [76] data.table_1.17.0        hms_1.1.3                xml2_1.3.6              </span></span>
<span><span class="co">##  [79] XVector_0.46.0           foreach_1.5.2            pillar_1.10.1           </span></span>
<span><span class="co">##  [82] stringr_1.5.1            circlize_0.4.16          dplyr_1.1.4             </span></span>
<span><span class="co">##  [85] BiocFileCache_2.14.0     lattice_0.22-6           deldir_2.0-4            </span></span>
<span><span class="co">##  [88] rtracklayer_1.66.0       bit_4.5.0.1              biovizBase_1.54.0       </span></span>
<span><span class="co">##  [91] tidyselect_1.2.1         locfit_1.5-9.11          pbapply_1.7-2           </span></span>
<span><span class="co">##  [94] Biostrings_2.74.1        knitr_1.49               gridExtra_2.3           </span></span>
<span><span class="co">##  [97] bookdown_0.42            ProtGenerics_1.38.0      xfun_0.51               </span></span>
<span><span class="co">## [100] stringi_1.8.4            UCSC.utils_1.2.0         lazyeval_0.2.2          </span></span>
<span><span class="co">## [103] yaml_2.3.10              evaluate_1.0.3           codetools_0.2-20        </span></span>
<span><span class="co">## [106] interp_1.1-6             GenomicFiles_1.42.0      tibble_3.2.1            </span></span>
<span><span class="co">## [109] BiocManager_1.30.25      cli_3.6.4                rpart_4.1.23            </span></span>
<span><span class="co">## [112] systemfonts_1.2.1        munsell_0.5.1            jquerylib_0.1.4         </span></span>
<span><span class="co">## [115] dichromat_2.0-0.1        Rcpp_1.0.14              dbplyr_2.5.0            </span></span>
<span><span class="co">## [118] png_0.1-8                XML_3.99-0.18            parallel_4.4.2          </span></span>
<span><span class="co">## [121] pkgdown_2.1.1            ggplot2_3.5.1            blob_1.2.4              </span></span>
<span><span class="co">## [124] prettyunits_1.2.0        jpeg_0.1-10              latticeExtra_0.6-30     </span></span>
<span><span class="co">## [127] AnnotationFilter_1.30.0  bitops_1.0-9             viridisLite_0.4.2       </span></span>
<span><span class="co">## [130] VariantAnnotation_1.52.0 scales_1.3.0             crayon_1.5.3            </span></span>
<span><span class="co">## [133] GetoptLong_1.0.5         rlang_1.1.5              cowplot_1.1.3           </span></span>
<span><span class="co">## [136] KEGGREST_1.46.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre-Luc Germain, Emanuel Sonder.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
