<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="epiwraps">
<title>Visualizing signals across many regions â€¢ epiwraps</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Visualizing signals across many regions">
<meta property="og:description" content="epiwraps">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">epiwraps</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.82</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/bam2bw.html">Generating bigwig tracks with bam2bw</a>
    <a class="dropdown-item" href="../articles/misc.html">Miscellaneous epiwraps functions</a>
    <a class="dropdown-item" href="../articles/multiRegionPlot.html">Visualizing signals across many regions</a>
    <a class="dropdown-item" href="../articles/singleRegionPlot.html">Visualizing signals in a single region</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Visualizing signals across many regions</h1>
                        <h4 data-toc-skip class="author">Pierre-Luc
Germain</h4>
            <address class="author_afil">
      D-HEST Institute for Neuroscience, ETHLab of Statistical
Bioinformatics, UZH<br><div class="d-none name"><code>multiRegionPlot.Rmd</code></div>
    </address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>This vignette covers the functions necessary for plotting
      signal across multiple regions. This involves acquiring positional
      information in/around given regions across tracks, functions to
      manipulate and aggregate these matrices, as well as functions to
      plot signal heatmaps from them.</p>
    </div>
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Since reading data from many regions is typically longer than
plotting it, we split plotting and acquiring the data. The latter is
done through function specific to this package, while the former
involves especially a wrapper around the <em><a href="https://bioconductor.org/packages/3.18/EnrichedHeatmap" class="external-link">EnrichedHeatmap</a></em>
package. The interface here has been simplified, but for full
functionality and customization it is recommended to have a look at the
<em><a href="https://bioconductor.org/packages/3.18/EnrichedHeatmap" class="external-link">EnrichedHeatmap</a></em>
documentation.</p>
<div class="section level3">
<h3 id="reading-signal-around-a-set-of-regions">Reading signal around a set of regions<a class="anchor" aria-label="anchor" href="#reading-signal-around-a-set-of-regions"></a>
</h3>
<p>The <code>signal2Matrix</code> function reads genomic signals around
the centers of a set of regions. It can read from bam and BigWig files,
although reading from bam files is considerably slower and we strongly
recommend using bigwig files. For generating bigwig files that show the
kind of signal you want to visualize, see the
<a href="bam2bw.html">vignette to this effect</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ethz-ins.github.io/epiwraps/">epiwraps</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># we fetch the path to the example bigwig file:</span></span>
<span><span class="va">bwf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example_atac.bw"</span>, package<span class="op">=</span><span class="st">"epiwraps"</span><span class="op">)</span></span>
<span><span class="co"># we load example regions (could be a GRanges or a path to a bed-like file):</span></span>
<span><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example_peaks.bed"</span>, package<span class="op">=</span><span class="st">"epiwraps"</span><span class="op">)</span></span>
<span><span class="co"># we obtain the matrix of the signal around the regions. For the purpose of this</span></span>
<span><span class="co"># example, we'll read twice from the same:</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/signal2Matrix.html">signal2Matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>atac1<span class="op">=</span><span class="va">bwf</span>, atac2<span class="op">=</span><span class="va">bwf</span><span class="op">)</span>, <span class="va">regions</span>, extend<span class="op">=</span><span class="fl">1000L</span>, binMethod <span class="op">=</span> <span class="st">"mean"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Reading /home/runner/work/_temp/Library/epiwraps/extdata/example_atac.bw</span></span>
<span><span class="co">## Reading /home/runner/work/_temp/Library/epiwraps/extdata/example_atac.bw</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span></span></code></pre></div>
<pre><code><span><span class="co">## class: EnrichmentSE </span></span>
<span><span class="co">## 2 tracks across 264 regions</span></span>
<span><span class="co">## assays(2): input enriched_score</span></span>
<span><span class="co">## rownames(264): 1:4327751 1:4327924 ... 1:194964294 1:195054176</span></span>
<span><span class="co">## rowData names(2): name score</span></span>
<span><span class="co">## colnames(2): atac1 atac2</span></span>
<span><span class="co">## colData names(0):</span></span>
<span><span class="co">## metadata(0):</span></span></code></pre>
<p>The result is an object of class <code>EnrichmentSE</code>, which
inherits from a <em><a href="https://bioconductor.org/packages/3.18/SummarizedExperiment/vignettes/RangedSummarizedExperiment" class="external-link">SummarizedExperiment</a></em>,
and therefore affords all the manipulations that the latter offers. Each
region is stored as a row, and each sample or signal track as a column
of the object. So we can see that we have signal for 264 rows/regions
from two tracks.</p>
<p>We could subset to the first 50 regions as follows:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">50</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## class: EnrichmentSE </span></span>
<span><span class="co">## 2 tracks across 50 regions</span></span>
<span><span class="co">## assays(2): input enriched_score</span></span>
<span><span class="co">## rownames(50): 1:4327751 1:4327924 ... 1:32002589 1:32002740</span></span>
<span><span class="co">## rowData names(2): name score</span></span>
<span><span class="co">## colnames(2): atac1 atac2</span></span>
<span><span class="co">## colData names(0):</span></span>
<span><span class="co">## metadata(0):</span></span></code></pre>
<p>or obtain the coordinates of the queried regions :</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 264 ranges and 2 metadata columns:</span></span>
<span><span class="co">##               seqnames    ranges strand |        name     score</span></span>
<span><span class="co">##                  &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##     1:4327751        1   4327751      * |        &lt;NA&gt;         0</span></span>
<span><span class="co">##     1:4327924        1   4327924      * |        &lt;NA&gt;         0</span></span>
<span><span class="co">##     1:4328528        1   4328528      * |        &lt;NA&gt;         0</span></span>
<span><span class="co">##     1:4927873        1   4927873      * |        &lt;NA&gt;         0</span></span>
<span><span class="co">##     1:5180899        1   5180899      * |        &lt;NA&gt;         0</span></span>
<span><span class="co">##           ...      ...       ...    ... .         ...       ...</span></span>
<span><span class="co">##   1:191885106        1 191885106      * |        &lt;NA&gt;         0</span></span>
<span><span class="co">##   1:191912950        1 191912950      * |        &lt;NA&gt;         0</span></span>
<span><span class="co">##   1:191937592        1 191937592      * |        &lt;NA&gt;         0</span></span>
<span><span class="co">##   1:194964294        1 194964294      * |        &lt;NA&gt;         0</span></span>
<span><span class="co">##   1:195054176        1 195054176      * |        &lt;NA&gt;         0</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>One can further obtain more detailed information about the bins saved
in the object:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/showTrackInfo.html">showTrackInfo</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## atac1 ( 264x200 ) :</span></span>
<span><span class="co">##   -1000/+1000bp (100 windows each)</span></span>
<span><span class="co">##   around the centers of given regions </span></span>
<span><span class="co">## atac2 ( 264x200 ) :</span></span>
<span><span class="co">##   -1000/+1000bp (100 windows each)</span></span>
<span><span class="co">##   around the centers of given regions</span></span></code></pre>
<p>This means that each signal track is a matrix of 200 columns, because
we asked to extend 1000bp on either side, and the default bin size is
10bp, making 100 bins/windows on each side.</p>
<div class="section level4">
<h4 id="extracting-and-manipulating-signal-matrices">Extracting and manipulating signal matrices<a class="anchor" aria-label="anchor" href="#extracting-and-manipulating-signal-matrices"></a>
</h4>
<p>It is possible to extract the list of signal matrix for
manipulations, e.g.Â for transformation:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># square-root transform</span></span>
<span><span class="va">m2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="../reference/getSignalMatrices.html">getSignalMatrices</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span>, <span class="va">sqrt</span><span class="op">)</span></span></code></pre></div>
<p>See <code><a href="../reference/addAssayToESE.html">?addAssayToESE</a></code> for adding a list of signal matrices
(such as <code>m2</code> here) to an existing <code>EnrichmentSE</code>
object. In addition, signal matrices can be combined, either manually or
using <code><a href="../reference/mergeSignalMatrices.html">?mergeSignalMatrices</a></code>.</p>
</div>
</div>
<div class="section level3">
<h3 id="plotting-heatmaps">Plotting heatmaps<a class="anchor" aria-label="anchor" href="#plotting-heatmaps"></a>
</h3>
<p>Once the signal has been read and the object prepared, (and
eventually normalized, see the section below), we can plot heatmaps
based on them as follows:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span></code></pre></div>
<p><img src="multiRegionPlot_files/figure-html/heatmap1-1.png" width="384">
We can use most arguments that are supported by <em><a href="https://bioconductor.org/packages/3.18/EnrichedHeatmap" class="external-link">EnrichedHeatmap</a></em>
(and thus, by extension, by <em><a href="https://bioconductor.org/packages/3.18/ComplexHeatmap" class="external-link">ComplexHeatmap</a></em>),
for example:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">m</span>, colors<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"white"</span>,<span class="st">"darkred"</span><span class="op">)</span>, cluster_rows<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                     show_row_dend<span class="op">=</span><span class="cn">TRUE</span>, top_annotation<span class="op">=</span><span class="cn">FALSE</span>, </span>
<span>                     row_title<span class="op">=</span><span class="st">"My list of cool regions"</span><span class="op">)</span></span></code></pre></div>
<p><img src="multiRegionPlot_files/figure-html/heatmap2-1.png" width="384"></p>
<div class="section level5">
<h5 id="color-scale-trimming">Color-scale trimming<a class="anchor" aria-label="anchor" href="#color-scale-trimming"></a>
</h5>
<p>By default, the colorscale is trimmed to prevent most of it being
driven by rare extreme values. This can be controlled via the
<code>trim</code> argument (which indicates up to which quantile of data
points to keep to establish the colorscale). Compare for instance the
following two heatmaps:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">m</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, trim<span class="op">=</span><span class="fl">1</span>, scale_title<span class="op">=</span><span class="st">"trim=1"</span>, column_title<span class="op">=</span><span class="st">"trim=1 (no trim)"</span>,</span>
<span>                     top_annotation<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">m</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, trim<span class="op">=</span><span class="fl">0.99</span>, scale_title<span class="op">=</span><span class="st">"trim=0.99"</span>,</span>
<span>                       column_title<span class="op">=</span><span class="st">"trim=0.99"</span>, top_annotation<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">m</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, trim<span class="op">=</span><span class="fl">0.9</span>, column_title<span class="op">=</span><span class="st">"trim=0.9"</span>,</span>
<span>                       scale_title<span class="op">=</span><span class="st">"trim=0.9"</span>, top_annotation<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="multiRegionPlot_files/figure-html/heatmapTrim-1.png" width="480"></p>
<p>The underlying data is exactly the same, only the color-mapping
changes. In the left one, which has no trimming, a single very high
value at the top forces the colorscale to extend to high values, even
though most of the data is in the very low range, resulting in a very
dark heatmap. In the one on the right, itâ€™s the opposite: so much is
trimmed that many points reach the top of the colorscale, resulting in a
an â€˜over-exposedâ€™ heatmap. In practice, it is advisable to use minimal
trimming (e.g.Â the default is <code>c(0.02,0.98)</code>).</p>
</div>
</div>
<div class="section level3">
<h3 id="normalization">Normalization<a class="anchor" aria-label="anchor" href="#normalization"></a>
</h3>
<p>We distinguish between two ways of calculating normalization factors:
either from the signal files (e.g.Â bigwig tracks), which is the most
robust way, or from an existing signal matrices (as for instance
contained in the above<br><code>EnrichmentSE</code> object).</p>
<div class="section level4">
<h4 id="obtaining-normalization-factors-for-a-set-of-signal-files">Obtaining normalization factors for a set of signal files<a class="anchor" aria-label="anchor" href="#obtaining-normalization-factors-for-a-set-of-signal-files"></a>
</h4>
<p>The <code><a href="../reference/getNormFactors.html">getNormFactors()</a></code> function can be used to estimate
normalization factors using different methods, such as background
normalization (aka SSN), shared-peaks (aka MAnorm), TMM, S3norm, etc.,
on a set of either i) bam files or ii) bigwig files.</p>
<p>The normalization factors can then be applied on signal matrices, for
example:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bwfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>atac1<span class="op">=</span><span class="va">bwf</span>, atac2<span class="op">=</span><span class="va">bwf</span><span class="op">)</span></span>
<span><span class="va">nf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getNormFactors.html">getNormFactors</a></span><span class="op">(</span><span class="va">bwfiles</span>, method<span class="op">=</span><span class="st">"background"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Comparing coverage in random regions...</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># in this case since the files are identical the normalization factors are both 1:</span></span>
<span><span class="va">nf</span></span></code></pre></div>
<pre><code><span><span class="co">## atac1 atac2 </span></span>
<span><span class="co">##     1     1</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/signal2Matrix.html">signal2Matrix</a></span><span class="op">(</span><span class="va">bwfiles</span>, <span class="va">regions</span>, extend<span class="op">=</span><span class="fl">1000L</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Reading /home/runner/work/_temp/Library/epiwraps/extdata/example_atac.bw</span></span></code></pre>
<pre><code><span><span class="co">## Reading /home/runner/work/_temp/Library/epiwraps/extdata/example_atac.bw</span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renormalizeSignalMatrices.html">renormalizeSignalMatrices</a></span><span class="op">(</span><span class="va">sm</span>, scaleFactors<span class="op">=</span><span class="va">nf</span><span class="op">)</span></span>
<span><span class="va">sm</span></span></code></pre></div>
<pre><code><span><span class="co">## class: EnrichmentSE </span></span>
<span><span class="co">## 2 tracks across 264 regions</span></span>
<span><span class="co">## assays(3): normalized input enriched_score</span></span>
<span><span class="co">## rownames(264): 1:4327751 1:4327924 ... 1:194964294 1:195054176</span></span>
<span><span class="co">## rowData names(2): name score</span></span>
<span><span class="co">## colnames(2): atac1 atac2</span></span>
<span><span class="co">## colData names(0):</span></span>
<span><span class="co">## metadata(0):</span></span></code></pre>
<p>The object now has a new assay, called <code>normalized</code>, which
has been put in front and therefore will be used for most downstream
usages.</p>
</div>
<div class="section level4">
<h4 id="normalization-of-the-signal-matrices-themselves">Normalization of the signal matrices themselves<a class="anchor" aria-label="anchor" href="#normalization-of-the-signal-matrices-themselves"></a>
</h4>
<p>It is also possible to normalize the signal matrices using factors
derived from the matrices themselves, using the
<code>renormalizeSignalMatrices</code> function. Specifically, beyond
providing scaling factors, two methods are available:</p>
<ul>
<li>
<code>method="border"</code> works on the assumption that the
left/right borders of the matrices represent background signal which
should be equal across samples. As a result, it will work only if 1) the
left/right borders of the matrices are sufficiently far from the signal
(e.g.Â peaks) to be chiefly noise, and 2) the signal-to-noise ratio is
comparable across tracks/samples.</li>
<li>
<code>method="top"</code> instead works on the assumption that the
highest signal (after some trimming of the extremes) should be the same
across tracks/samples.</li>
</ul>
<p>To illustrate these, we will first introduce some difference between
our two tracks using arbitrary factors:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renormalizeSignalMatrices.html">renormalizeSignalMatrices</a></span><span class="op">(</span><span class="va">sm</span>, scaleFactors<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">4</span><span class="op">)</span>, toAssay<span class="op">=</span><span class="st">"test"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">sm</span>, assay <span class="op">=</span> <span class="st">"test"</span><span class="op">)</span></span></code></pre></div>
<p><img src="multiRegionPlot_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>Then we can normalize:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renormalizeSignalMatrices.html">renormalizeSignalMatrices</a></span><span class="op">(</span><span class="va">sm</span>, method<span class="op">=</span><span class="st">"top"</span>, fromAssay<span class="op">=</span><span class="st">"test"</span><span class="op">)</span></span>
<span><span class="co"># again this adds an assay to the object, which will be automatically used when plotting:</span></span>
<span><span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">sm</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Using assay topNormalized</span></span></code></pre>
<p><img src="multiRegionPlot_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>And weâ€™ve recovered comparable signal across the two
tracks/samples.</p>
</div>
<div class="section level4">
<h4 id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h4>
<p>The traditional ranking by decreasing overall enrichment can easily
hide patterns in the data, which are instead revealed by clustering.
While hierarchical clustering can be performed in the plotting function
itself, this is based on the whole enrichment profiles and seldom
provides good results in practice. For this reason, we provide a
function that clusters the rows based on the distance-weighted
enrichment scores. For example:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># not run - would fail here due to the absence of variability between tracks</span></span>
<span><span class="va">cl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clusterSignalMatrices.html">clusterSignalMatrices</a></span><span class="op">(</span><span class="va">sm</span>, k<span class="op">=</span><span class="fl">3</span>, scaleRows<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># we could store this in the rowData of the object:</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sm</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="va">cl</span></span>
<span><span class="va">mycolors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1"</span><span class="op">=</span><span class="st">"red"</span>, <span class="st">"2"</span><span class="op">=</span><span class="st">"blue"</span>, <span class="st">"3"</span><span class="op">=</span><span class="st">"darkgreen"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">sm</span>, row_split<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sm</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span>, mean_color<span class="op">=</span><span class="va">mycolors</span><span class="op">)</span></span></code></pre></div>
<p>Note that here we are splitting into 3 clusters, you can also provide
a range of values (e.g.Â <code>k=2:8</code>) and the function will also
return cluster quality metrics for each.</p>
<p>[include proper exampleâ€¦ Creb family]</p>
</div>
</div>
<div class="section level3">
<h3 id="plotting-aggregated-signals">Plotting aggregated signals<a class="anchor" aria-label="anchor" href="#plotting-aggregated-signals"></a>
</h3>
<p>It is also possible to plot only the average signals across regions.
To do this, we first melt the signal matrices and then use <em><a href="https://CRAN.R-project.org/package=ggplot2" class="external-link">ggplot2</a></em>. The
<code>meltSignals</code> function will return a data.frame showing the
mean, standard deviation, standard error and median at each position
relative to the center, for each sample/matrix:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/meltSignals.html">meltSignals</a></span><span class="op">(</span><span class="va">sm</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Using assay topNormalized</span></span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">d</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   position sample      mean        SD          SE    median</span></span>
<span><span class="co">## 1    -1000  atac1 0.1428603 0.1100839 0.006775199 0.1222137</span></span>
<span><span class="co">## 2     -990  atac1 0.1433233 0.1173983 0.007225367 0.1222137</span></span>
<span><span class="co">## 3     -980  atac1 0.1444343 0.1173860 0.007224609 0.1222137</span></span>
<span><span class="co">## 4     -970  atac1 0.1455453 0.1180579 0.007265960 0.1222137</span></span>
<span><span class="co">## 5     -960  atac1 0.1451750 0.1190121 0.007324686 0.1222137</span></span>
<span><span class="co">## 6     -950  atac1 0.1467490 0.1182650 0.007278705 0.1222137</span></span></code></pre>
<p>This can then be used for plotting. For this example, since it
consists of twice the same file, weâ€™ll first create a fake
difference:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create a fake difference so that there's something to visualize in this example:</span></span>
<span><span class="va">d</span><span class="op">$</span><span class="va">mean</span><span class="op">[</span><span class="va">d</span><span class="op">$</span><span class="va">sample</span><span class="op">==</span><span class="st">"atac2"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0.06</span><span class="op">+</span><span class="op">(</span><span class="va">d</span><span class="op">$</span><span class="va">mean</span><span class="op">[</span><span class="va">d</span><span class="op">$</span><span class="va">sample</span><span class="op">==</span><span class="st">"atac2"</span><span class="op">]</span><span class="op">-</span><span class="fl">0.06</span><span class="op">)</span><span class="op">*</span><span class="fl">0.6</span></span></code></pre></div>
<p>Then we can plot normally with <code>ggplot</code>:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">d</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">position</span>, <span class="va">mean</span>, colour<span class="op">=</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept<span class="op">=</span><span class="fl">0</span>, linetype<span class="op">=</span><span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_ribbon.html" class="external-link">geom_ribbon</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">position</span>, ymin<span class="op">=</span><span class="va">mean</span><span class="op">-</span><span class="va">SE</span>, ymax<span class="op">=</span><span class="va">mean</span><span class="op">+</span><span class="va">SE</span>, fill<span class="op">=</span><span class="va">sample</span><span class="op">)</span>, alpha<span class="op">=</span><span class="fl">0.4</span>, colour<span class="op">=</span><span class="cn">NA</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">1.2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"relative position"</span>, y<span class="op">=</span><span class="st">"mean RPKM"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">â„¹</span> Please use `linewidth` instead.</span></span>
<span><span class="co">## <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">## <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">## <span style="color: #555555;">generated.</span></span></span></code></pre>
<p><img src="multiRegionPlot_files/figure-html/aggPlot-1.png" width="384"></p>
<p>Note that when the regions are clustered, the clusters can be passed
to melt and aggregate separately the different clusters, e.g.:
<code>d &lt;- meltSignals(sm, splitBy=cl)</code>.</p>
<p><br><br></p>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.3 (2024-02-29)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] ggplot2_3.5.0               epiwraps_0.99.82           </span></span>
<span><span class="co">##  [3] EnrichedHeatmap_1.32.0      ComplexHeatmap_2.18.0      </span></span>
<span><span class="co">##  [5] SummarizedExperiment_1.32.0 Biobase_2.62.0             </span></span>
<span><span class="co">##  [7] GenomicRanges_1.54.1        GenomeInfoDb_1.38.7        </span></span>
<span><span class="co">##  [9] IRanges_2.36.0              S4Vectors_0.40.2           </span></span>
<span><span class="co">## [11] BiocGenerics_0.48.1         MatrixGenerics_1.14.0      </span></span>
<span><span class="co">## [13] matrixStats_1.2.0           BiocStyle_2.30.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3       rstudioapi_0.15.0        jsonlite_1.8.8          </span></span>
<span><span class="co">##   [4] shape_1.4.6.1            magrittr_2.0.3           GenomicFeatures_1.54.3  </span></span>
<span><span class="co">##   [7] farver_2.1.1             rmarkdown_2.26           GlobalOptions_0.1.2     </span></span>
<span><span class="co">##  [10] fs_1.6.3                 BiocIO_1.12.0            zlibbioc_1.48.0         </span></span>
<span><span class="co">##  [13] ragg_1.2.7               vctrs_0.6.5              memoise_2.0.1           </span></span>
<span><span class="co">##  [16] Rsamtools_2.18.0         RCurl_1.98-1.14          base64enc_0.1-3         </span></span>
<span><span class="co">##  [19] htmltools_0.5.7          S4Arrays_1.2.1           progress_1.2.3          </span></span>
<span><span class="co">##  [22] curl_5.2.1               SparseArray_1.2.4        Formula_1.2-5           </span></span>
<span><span class="co">##  [25] sass_0.4.8               bslib_0.6.1              htmlwidgets_1.6.4       </span></span>
<span><span class="co">##  [28] desc_1.4.3               plyr_1.8.9               Gviz_1.46.1             </span></span>
<span><span class="co">##  [31] cachem_1.0.8             GenomicAlignments_1.38.2 lifecycle_1.0.4         </span></span>
<span><span class="co">##  [34] iterators_1.0.14         pkgconfig_2.0.3          Matrix_1.6-5            </span></span>
<span><span class="co">##  [37] R6_2.5.1                 fastmap_1.1.1            GenomeInfoDbData_1.2.11 </span></span>
<span><span class="co">##  [40] clue_0.3-65              digest_0.6.34            colorspace_2.1-0        </span></span>
<span><span class="co">##  [43] AnnotationDbi_1.64.1     textshaping_0.3.7        Hmisc_5.1-1             </span></span>
<span><span class="co">##  [46] RSQLite_2.3.5            labeling_0.4.3           filelock_1.0.3          </span></span>
<span><span class="co">##  [49] fansi_1.0.6              httr_1.4.7               abind_1.4-5             </span></span>
<span><span class="co">##  [52] compiler_4.3.3           withr_3.0.0              bit64_4.0.5             </span></span>
<span><span class="co">##  [55] doParallel_1.0.17        backports_1.4.1          htmlTable_2.4.2         </span></span>
<span><span class="co">##  [58] BiocParallel_1.36.0      DBI_1.2.2                UpSetR_1.4.0            </span></span>
<span><span class="co">##  [61] highr_0.10               biomaRt_2.58.2           rappdirs_0.3.3          </span></span>
<span><span class="co">##  [64] DelayedArray_0.28.0      rjson_0.2.21             tools_4.3.3             </span></span>
<span><span class="co">##  [67] foreign_0.8-86           nnet_7.3-19              glue_1.7.0              </span></span>
<span><span class="co">##  [70] restfulr_0.0.15          checkmate_2.3.1          cluster_2.1.6           </span></span>
<span><span class="co">##  [73] generics_0.1.3           gtable_0.3.4             BSgenome_1.70.2         </span></span>
<span><span class="co">##  [76] ensembldb_2.26.0         data.table_1.15.2        hms_1.1.3               </span></span>
<span><span class="co">##  [79] xml2_1.3.6               utf8_1.2.4               XVector_0.42.0          </span></span>
<span><span class="co">##  [82] foreach_1.5.2            pillar_1.9.0             stringr_1.5.1           </span></span>
<span><span class="co">##  [85] circlize_0.4.16          dplyr_1.1.4              BiocFileCache_2.10.1    </span></span>
<span><span class="co">##  [88] lattice_0.22-5           deldir_2.0-4             rtracklayer_1.62.0      </span></span>
<span><span class="co">##  [91] bit_4.0.5                biovizBase_1.50.0        tidyselect_1.2.0        </span></span>
<span><span class="co">##  [94] locfit_1.5-9.9           pbapply_1.7-2            Biostrings_2.70.2       </span></span>
<span><span class="co">##  [97] knitr_1.45               gridExtra_2.3            bookdown_0.38           </span></span>
<span><span class="co">## [100] ProtGenerics_1.34.0      xfun_0.42                stringi_1.8.3           </span></span>
<span><span class="co">## [103] lazyeval_0.2.2           yaml_2.3.8               evaluate_0.23           </span></span>
<span><span class="co">## [106] codetools_0.2-19         interp_1.1-6             GenomicFiles_1.38.0     </span></span>
<span><span class="co">## [109] tibble_3.2.1             BiocManager_1.30.22      cli_3.6.2               </span></span>
<span><span class="co">## [112] rpart_4.1.23             systemfonts_1.0.5        munsell_0.5.0           </span></span>
<span><span class="co">## [115] jquerylib_0.1.4          dichromat_2.0-0.1        Rcpp_1.0.12             </span></span>
<span><span class="co">## [118] dbplyr_2.4.0             png_0.1-8                XML_3.99-0.16.1         </span></span>
<span><span class="co">## [121] parallel_4.3.3           pkgdown_2.0.7            blob_1.2.4              </span></span>
<span><span class="co">## [124] prettyunits_1.2.0        jpeg_0.1-10              latticeExtra_0.6-30     </span></span>
<span><span class="co">## [127] AnnotationFilter_1.26.0  bitops_1.0-7             viridisLite_0.4.2       </span></span>
<span><span class="co">## [130] VariantAnnotation_1.48.1 scales_1.3.0             purrr_1.0.2             </span></span>
<span><span class="co">## [133] crayon_1.5.2             GetoptLong_1.0.5         rlang_1.1.3             </span></span>
<span><span class="co">## [136] cowplot_1.1.3            KEGGREST_1.42.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Pierre-Luc Germain, Emanuel Sonder.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
