<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="epiwraps">
<title>Visualizing signals across many regions • epiwraps</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Visualizing signals across many regions">
<meta property="og:description" content="epiwraps">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">epiwraps</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.95</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/bam2bw.html">Generating bigwig tracks with bam2bw</a>
    <a class="dropdown-item" href="../articles/misc.html">Miscellaneous epiwraps functions</a>
    <a class="dropdown-item" href="../articles/multiRegionPlot.html">Visualizing signals across many regions</a>
    <a class="dropdown-item" href="../articles/normalization.html">Normalizing genomic signals</a>
    <a class="dropdown-item" href="../articles/singleRegionPlot.html">Visualizing signals in a single region</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Visualizing signals across many regions</h1>
                        <h4 data-toc-skip class="author">Pierre-Luc
Germain</h4>
            <address class="author_afil">
      Lab of Statistical Bioinformatics, University of Zürich;D-HEST
Institute for Neuroscience, ETH Zürich,
Switzerland<br><div class="d-none name"><code>multiRegionPlot.Rmd</code></div>
    </address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>This vignette covers the functions necessary for plotting
      signal across multiple regions. This involves acquiring positional
      information in/around given regions across tracks, functions to
      manipulate and aggregate these matrices, as well as functions to
      plot signal heatmaps from them.</p>
    </div>
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Since reading data from many regions is typically longer than
plotting it, we split plotting and acquiring the data. The latter is
done through function specific to this package, while the former wraps
around the <em><a href="https://bioconductor.org/packages/3.19/EnrichedHeatmap" class="external-link">EnrichedHeatmap</a></em>
package. The interface here has been simplified, but for full
functionality and customization it is recommended to have a look at the
<em><a href="https://bioconductor.org/packages/3.19/EnrichedHeatmap" class="external-link">EnrichedHeatmap</a></em>
documentation.</p>
</div>
<div class="section level2">
<h2 id="reading-signal-inaround-a-set-of-regions">Reading signal in/around a set of regions<a class="anchor" aria-label="anchor" href="#reading-signal-inaround-a-set-of-regions"></a>
</h2>
<p>The <code>signal2Matrix</code> function reads genomic signals around
the centers of a set of regions. It can read from bam and BigWig files,
although reading from bam files is considerably slower and we strongly
recommend using bigwig files. For generating bigwig files that show the
kind of signal you want to visualize, see the
<a href="bam2bw.html">vignette to this effect</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ethz-ins.github.io/epiwraps/">epiwraps</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># we fetch the path to the example bigwig file:</span></span>
<span><span class="va">bwf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example_atac.bw"</span>, package<span class="op">=</span><span class="st">"epiwraps"</span><span class="op">)</span></span>
<span><span class="co"># we load example regions (could be a GRanges or a path to a bed-like file):</span></span>
<span><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example_peaks.bed"</span>, package<span class="op">=</span><span class="st">"epiwraps"</span><span class="op">)</span></span>
<span><span class="co"># we obtain the matrix of the signal around the regions. For the purpose of this</span></span>
<span><span class="co"># example, we'll read twice from the same:</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/signal2Matrix.html">signal2Matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>atac1<span class="op">=</span><span class="va">bwf</span>, atac2<span class="op">=</span><span class="va">bwf</span><span class="op">)</span>, <span class="va">regions</span>, extend<span class="op">=</span><span class="fl">1000L</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Reading /home/runner/work/_temp/Library/epiwraps/extdata/example_atac.bw</span></span>
<span><span class="co">## Reading /home/runner/work/_temp/Library/epiwraps/extdata/example_atac.bw</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span></span></code></pre></div>
<pre><code><span><span class="co">## class: EnrichmentSE </span></span>
<span><span class="co">## 2 tracks across 150 regions</span></span>
<span><span class="co">## assays(1): input</span></span>
<span><span class="co">## rownames(150): 1:195054101-195054250 1:133522798-133523047 ...</span></span>
<span><span class="co">##   1:22224734-22224983 1:90375438-90375787</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(2): atac1 atac2</span></span>
<span><span class="co">## colData names(0):</span></span>
<span><span class="co">## metadata(0):</span></span></code></pre>
<p>The result is an object of class <code>EnrichmentSE</code>, which
inherits from a <em><a href="https://bioconductor.org/packages/3.19/SummarizedExperiment/vignettes/RangedSummarizedExperiment" class="external-link">SummarizedExperiment</a></em>,
and therefore affords all the manipulations that the latter offers. Each
region is stored as a row, and each sample or signal track as a column
of the object. So we can see that we have signal for 150 rows/regions
from two tracks.</p>
<p>We could subset to the first 50 regions as follows:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">50</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## class: EnrichmentSE </span></span>
<span><span class="co">## 2 tracks across 50 regions</span></span>
<span><span class="co">## assays(1): input</span></span>
<span><span class="co">## rownames(50): 1:195054101-195054250 1:133522798-133523047 ...</span></span>
<span><span class="co">##   1:16142722-16143021 1:143652894-143653043</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(2): atac1 atac2</span></span>
<span><span class="co">## colData names(0):</span></span>
<span><span class="co">## metadata(0):</span></span></code></pre>
<p>or obtain the coordinates of the queried regions :</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 150 ranges and 0 metadata columns:</span></span>
<span><span class="co">##                         seqnames              ranges strand</span></span>
<span><span class="co">##                            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   1:195054101-195054250        1 195054101-195054250      *</span></span>
<span><span class="co">##   1:133522798-133523047        1 133522798-133523047      *</span></span>
<span><span class="co">##   1:133522621-133522870        1 133522621-133522870      *</span></span>
<span><span class="co">##   1:170892151-170892250        1 170892151-170892250      *</span></span>
<span><span class="co">##   1:170897239-170897438        1 170897239-170897438      *</span></span>
<span><span class="co">##                     ...      ...                 ...    ...</span></span>
<span><span class="co">##   1:121131630-121131729        1 121131630-121131729      *</span></span>
<span><span class="co">##   1:122700285-122700334        1 122700285-122700334      *</span></span>
<span><span class="co">##   1:100107921-100108020        1 100107921-100108020      *</span></span>
<span><span class="co">##     1:22224734-22224983        1   22224734-22224983      *</span></span>
<span><span class="co">##     1:90375438-90375787        1   90375438-90375787      *</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>One can further obtain more detailed information about the bins saved
in the object:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/showTrackInfo.html">showTrackInfo</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## atac1 ( 150x200 ) :</span></span>
<span><span class="co">##   -1kb/+1kb (100 windows each)</span></span>
<span><span class="co">##   around the centers of given regions </span></span>
<span><span class="co">## atac2 ( 150x200 ) :</span></span>
<span><span class="co">##   -1kb/+1kb (100 windows each)</span></span>
<span><span class="co">##   around the centers of given regions</span></span></code></pre>
<p>This means that each signal track is a matrix of 200 columns, because
we asked to extend 1000bp on either side, and the default bin size is
10bp, making 100 bins/windows on each side.</p>
<div class="section level3">
<h3 id="extracting-and-manipulating-signal-matrices">Extracting and manipulating signal matrices<a class="anchor" aria-label="anchor" href="#extracting-and-manipulating-signal-matrices"></a>
</h3>
<p>It is possible to extract the list of signal matrix for
manipulations, e.g. for transformation:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># square-root transform</span></span>
<span><span class="va">m2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="../reference/getSignalMatrices.html">getSignalMatrices</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span>, <span class="va">sqrt</span><span class="op">)</span></span></code></pre></div>
<p>See <code><a href="../reference/addAssayToESE.html">?addAssayToESE</a></code> for adding a list of signal matrices
(such as <code>m2</code> here) to an existing <code>EnrichmentSE</code>
object. In addition, signal matrices can be combined, either manually or
using <code><a href="../reference/mergeSignalMatrices.html">?mergeSignalMatrices</a></code>.</p>
</div>
<div class="section level3">
<h3 id="normalization">Normalization<a class="anchor" aria-label="anchor" href="#normalization"></a>
</h3>
<p>By default, bigwig files generated by <code>epiwraps</code> are
normalized for library size, but this is not always sufficient.
<code>EnrichmentSE</code> can further be normalized using various
methods. For an overview of these methods, see the
<a href="normalization.html">normalization vignette</a>.</p>
<p><br><br></p>
</div>
</div>
<div class="section level2">
<h2 id="plotting-heatmaps">Plotting heatmaps<a class="anchor" aria-label="anchor" href="#plotting-heatmaps"></a>
</h2>
<p>Once the signal has been read and the object prepared, (and
eventually normalized, see the section below), we can plot heatmaps
based on them as follows:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span></code></pre></div>
<p><img src="multiRegionPlot_files/figure-html/heatmap1-1.png" width="384">
We can use most arguments that are supported by <em><a href="https://bioconductor.org/packages/3.19/EnrichedHeatmap" class="external-link">EnrichedHeatmap</a></em>
(and thus, by extension, by <em><a href="https://bioconductor.org/packages/3.19/ComplexHeatmap" class="external-link">ComplexHeatmap</a></em>),
for example:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">m</span>, colors<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"white"</span>,<span class="st">"darkred"</span><span class="op">)</span>, cluster_rows<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                     show_row_dend<span class="op">=</span><span class="cn">TRUE</span>, top_annotation<span class="op">=</span><span class="cn">FALSE</span>, </span>
<span>                     row_title<span class="op">=</span><span class="st">"My list of cool regions"</span><span class="op">)</span></span></code></pre></div>
<p><img src="multiRegionPlot_files/figure-html/heatmap2-1.png" width="384"></p>
<p>It is often useful to subset to regions with a high enrichment, which
we can do with the <code>score</code> function:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">m</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/score.html" class="external-link">score</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,<span class="fl">50</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="multiRegionPlot_files/figure-html/heatmap2b-1.png" width="384"></p>
<div class="section level3">
<h3 id="color-scale-trimming">Color-scale trimming<a class="anchor" aria-label="anchor" href="#color-scale-trimming"></a>
</h3>
<p>By default, the colorscale is trimmed to prevent most of it being
driven by rare extreme values. This can be controlled via the
<code>trim</code> argument (which indicates up to which quantile of data
points to keep to establish the colorscale). Compare for instance the
following two heatmaps:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">m</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, trim<span class="op">=</span><span class="fl">1</span>, scale_title<span class="op">=</span><span class="st">"trim=1"</span>, column_title<span class="op">=</span><span class="st">"trim=1 (no trim)"</span>,</span>
<span>                     top_annotation<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">m</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, trim<span class="op">=</span><span class="fl">0.99</span>, scale_title<span class="op">=</span><span class="st">"trim=0.99"</span>,</span>
<span>                       column_title<span class="op">=</span><span class="st">"trim=0.99"</span>, top_annotation<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">m</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, trim<span class="op">=</span><span class="fl">0.9</span>, column_title<span class="op">=</span><span class="st">"trim=0.9"</span>,</span>
<span>                       scale_title<span class="op">=</span><span class="st">"trim=0.9"</span>, top_annotation<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="multiRegionPlot_files/figure-html/heatmapTrim-1.png" width="480"></p>
<p>The underlying data is exactly the same, only the color-mapping
changes. In the left one, which has no trimming, a single very high
value at the top forces the colorscale to extend to high values, even
though most of the data is in the very low range, resulting in a very
dark heatmap. In the one on the right, it’s the opposite: so much is
trimmed that many points reach the top of the colorscale, resulting in a
an ‘over-exposed’ heatmap. In practice, it is advisable to use minimal
trimming (e.g. the default is <code>c(0.02,0.98)</code>).</p>
</div>
<div class="section level3">
<h3 id="different-colorscales-for-different-tracks">Different colorscales for different tracks<a class="anchor" aria-label="anchor" href="#different-colorscales-for-different-tracks"></a>
</h3>
<p>It is also possible to have different colorscales for different
tracks, which is especially useful when comparing very different
signals. To illustrate this, let’s load an example with different
tracks:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">exampleESE</span><span class="op">)</span></span>
<span><span class="va">exampleESE</span></span></code></pre></div>
<pre><code><span><span class="co">## class: EnrichmentSE </span></span>
<span><span class="co">## 3 tracks across 150 regions</span></span>
<span><span class="co">## assays(1): input</span></span>
<span><span class="co">## rownames(150): chr1:36986026-36986320 chr1:36986855-36987064 ...</span></span>
<span><span class="co">##   chr1:135509406-135510907 chr1:131525035-131527379</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(3): H3K27ac H3K4me3 p300</span></span>
<span><span class="co">## colData names(0):</span></span>
<span><span class="co">## metadata(0):</span></span></code></pre>
<p>We can put each of the three tracks on its own color scale:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">exampleESE</span>, multiScale<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="multiRegionPlot_files/figure-html/heatmap1ms-1.png" width="432"></p>
<p>One could also specify colors separately by providing them as a
list:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">exampleESE</span>,</span>
<span>                     colors<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"white"</span>,<span class="st">"darkblue"</span><span class="op">)</span>, <span class="st">"darkgreen"</span>, <span class="st">"darkred"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="multiRegionPlot_files/figure-html/heatmap1ms2-1.png" width="432"></p>
<p>This information can also be stored in the object, rather than
specified everytime:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exampleESE</span><span class="op">$</span><span class="va">hmcolors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu">viridisLite</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">inferno</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,  <span class="st">"darkgreen"</span>, <span class="st">"darkred"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">exampleESE</span><span class="op">)</span></span></code></pre></div>
<p><img src="multiRegionPlot_files/figure-html/unnamed-chunk-8-1.png" width="432"></p>
</div>
<div class="section level3">
<h3 id="scaled-regions">Scaled regions<a class="anchor" aria-label="anchor" href="#scaled-regions"></a>
</h3>
<p>By default, <code>signal2Matrix</code> looks at a pre-defined
interval (defined by the <code>extend</code> argument) around the center
of the provided regions. This means that the width of the input regions
is ignored. In some circumstances, however, it can be useful to scale
regions to the same width, which can be done using the
<code>type="scaled"</code> argument. Consider the following example:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ese</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/signal2Matrix.html">signal2Matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>center<span class="op">=</span><span class="va">bwf</span><span class="op">)</span>, <span class="va">regions</span>, extend<span class="op">=</span><span class="fl">1000L</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/signal2Matrix.html">signal2Matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>scaled<span class="op">=</span><span class="va">bwf</span><span class="op">)</span>, <span class="va">regions</span>, extend<span class="op">=</span><span class="fl">1000L</span>, type<span class="op">=</span><span class="st">"scaled"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Reading /home/runner/work/_temp/Library/epiwraps/extdata/example_atac.bw</span></span>
<span><span class="co">## Reading /home/runner/work/_temp/Library/epiwraps/extdata/example_atac.bw</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">ese</span><span class="op">)</span></span></code></pre></div>
<p><img src="multiRegionPlot_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>For some purposes, such as when plotting signal over transcripts, it
can be useful for the target region to consist of multiple regions
(e.g. exons) stitched together. This can be done by supplying a
<code>GRangesList</code> as regions:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># we make a dummy GRangesList:</span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">import</span><span class="op">(</span><span class="va">regions</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dummy.grl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRangesList-class.html" class="external-link">GRangesList</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/splitAsList.html" class="external-link">split</a></span><span class="op">(</span><span class="va">a</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">LETTERS</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">]</span>,each<span class="op">=</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/signal2Matrix.html">signal2Matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>scaled<span class="op">=</span><span class="va">bwf</span><span class="op">)</span>, <span class="va">dummy.grl</span>, extend<span class="op">=</span><span class="fl">1000L</span>, type<span class="op">=</span><span class="st">"scaled"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Reading /home/runner/work/_temp/Library/epiwraps/extdata/example_atac.bw</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">sm</span><span class="op">)</span></span></code></pre></div>
<p><img src="multiRegionPlot_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="heatmap-rasterization">Heatmap rasterization<a class="anchor" aria-label="anchor" href="#heatmap-rasterization"></a>
</h3>
<p>When plotting more regions that there are pixels available, several
regions have to be summarized in one pixel, and doing this before
generating the heatmap makes the plot much less heavy.</p>
<p>By default, <code>EnrichedHeatmap</code> performs rasterization using
the <code>magick</code> package when it is installed, and falls back to
a very suboptimal method when not. It is therefore recommended to
install the <code>magick</code> package.</p>
<p>Depending on your settings, if the heatmap is <em>very big</em> you
might hit the preset limits of ‘magick’ base rasterization, which could
result in an error such as ‘Image must have at least 1 frame to write a
bitmap’. In such cases, you might have to degrade to a lower-quality
rasterization using the additional arguments
<code>raster_by_magick=FALSE, raster_device="CairoJPEG"</code>.</p>
<p>Finally, on some systems, the rasterization sometimes encounters an
‘UnableToOpenBlob’ error. At the moment, the only workaround this has
been to disable rasterization using <code>use_raster=FALSE</code>.</p>
</div>
</div>
<div class="section level2">
<h2 id="sorting-and-clustering">Sorting and clustering<a class="anchor" aria-label="anchor" href="#sorting-and-clustering"></a>
</h2>
<p>The traditional ranking by decreasing overall enrichment can easily
hide patterns in the data, which are instead revealed by clustering. One
approach is to use hierarchical clustering of the rows:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">exampleESE</span>, cluster_rows<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="multiRegionPlot_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>In this example, this already reveals important patterns in the data,
namely the fact that p300 binding is associated with H3K27ac and tends
to be mutually exclusive with the promoter-associated H3K4me3 mark.</p>
<p>The hierarchical clustering is based on the whole enrichment profile,
can easily be led astray by patterns in individual signals, and seldom
provides good results in practice. An alternative is to use enrichment
score weighted by distance to the center, eventually row-normalized, to
cluster the regions. We provide a function to this end:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># we cluster the regions using 2 clusters, and store the cluster labels in the </span></span>
<span><span class="co"># rowData of the object: </span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">exampleESE</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clusterSignalMatrices.html">clusterSignalMatrices</a></span><span class="op">(</span><span class="va">exampleESE</span>, k<span class="op">=</span><span class="fl">2</span>, scaleRows<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   ~95% of the variance explained by clusters</span></span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># we additionally label the clusters with colors:</span></span>
<span><span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">exampleESE</span>, row_split<span class="op">=</span><span class="st">"cluster"</span>,</span>
<span>                     mean_color<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1"</span><span class="op">=</span><span class="st">"red"</span>, <span class="st">"2"</span><span class="op">=</span><span class="st">"blue"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="multiRegionPlot_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>Note that here we are splitting into 3 clusters, you can also provide
a range of values (e.g. <code>k=2:8</code>) and the function will also
return cluster quality metrics for each.</p>
<p>Users can of course use their own algorithms to cluster regions. To
this end, it can be useful to summarize each region in each sample using
a single score which weights the signal according to the distance to the
center of the target (see
<code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/enriched_score.html" class="external-link">?EnrichedHeatmap::enriched_score</a></code>). This can be done for the
default assay with the <code>score</code> method:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/score.html" class="external-link">score</a></span><span class="op">(</span><span class="va">exampleESE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                           H3K27ac   H3K4me3     p300</span></span>
<span><span class="co">## chr1:36986026-36986320   39609.53   341.850 27443.12</span></span>
<span><span class="co">## chr1:36986855-36987064   48642.90   230.900 21184.08</span></span>
<span><span class="co">## chr1:36061123-36061559   26838.05   313.025 20699.60</span></span>
<span><span class="co">## chr1:36060725-36060951   34205.22   330.125 20061.05</span></span>
<span><span class="co">## chr1:180888607-180889217 20013.45  2019.725 22293.35</span></span>
<span><span class="co">## chr1:86487735-86488859   20008.00 26012.825  1265.85</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="plotting-aggregated-signals">Plotting aggregated signals<a class="anchor" aria-label="anchor" href="#plotting-aggregated-signals"></a>
</h2>
<p>It is also possible to plot only the average signals across regions.
To do this, we first melt the signal matrices and then use <em><a href="https://CRAN.R-project.org/package=ggplot2" class="external-link">ggplot2</a></em>. The
<code>meltSignals</code> function will return a data.frame showing the
mean, standard deviation, standard error and median at each position
relative to the center, for each sample/matrix:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/meltSignals.html">meltSignals</a></span><span class="op">(</span><span class="va">exampleESE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">d</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   position  sample     mean       SD       SE median</span></span>
<span><span class="co">## 1    -2000 H3K27ac 83.28000 182.5112 14.90197   11.5</span></span>
<span><span class="co">## 2    -1950 H3K27ac 85.87333 189.9628 15.51040   11.5</span></span>
<span><span class="co">## 3    -1900 H3K27ac 91.19333 207.2911 16.92525    9.0</span></span>
<span><span class="co">## 4    -1850 H3K27ac 90.92000 201.2272 16.43013    8.5</span></span>
<span><span class="co">## 5    -1800 H3K27ac 88.02667 192.6586 15.73051   10.0</span></span>
<span><span class="co">## 6    -1750 H3K27ac 85.52000 173.5301 14.16867   14.5</span></span></code></pre>
<p>This can then be used for plotting, simply with
<code>ggplot</code>:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">d</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">position</span>, <span class="va">mean</span>, colour<span class="op">=</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept<span class="op">=</span><span class="fl">0</span>, linetype<span class="op">=</span><span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_ribbon.html" class="external-link">geom_ribbon</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">position</span>, ymin<span class="op">=</span><span class="va">mean</span><span class="op">-</span><span class="va">SE</span>, ymax<span class="op">=</span><span class="va">mean</span><span class="op">+</span><span class="va">SE</span>, fill<span class="op">=</span><span class="va">sample</span><span class="op">)</span>, alpha<span class="op">=</span><span class="fl">0.4</span>, colour<span class="op">=</span><span class="cn">NA</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>linewidth<span class="op">=</span><span class="fl">1.2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"relative position"</span>, y<span class="op">=</span><span class="st">"mean RPKM"</span><span class="op">)</span></span></code></pre></div>
<p><img src="multiRegionPlot_files/figure-html/aggPlot-1.png" width="384"></p>
<p>We could also include cluster information:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/meltSignals.html">meltSignals</a></span><span class="op">(</span><span class="va">exampleESE</span>, splitBy <span class="op">=</span> <span class="st">"cluster"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">d</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">position</span>, <span class="va">mean</span>, colour<span class="op">=</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept<span class="op">=</span><span class="fl">0</span>, linetype<span class="op">=</span><span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_ribbon.html" class="external-link">geom_ribbon</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">position</span>, ymin<span class="op">=</span><span class="va">mean</span><span class="op">-</span><span class="va">SE</span>, ymax<span class="op">=</span><span class="va">mean</span><span class="op">+</span><span class="va">SE</span>, fill<span class="op">=</span><span class="va">sample</span><span class="op">)</span>, alpha<span class="op">=</span><span class="fl">0.4</span>, colour<span class="op">=</span><span class="cn">NA</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>linewidth<span class="op">=</span><span class="fl">1.2</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">split</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"relative position"</span>, y<span class="op">=</span><span class="st">"mean RPKM"</span><span class="op">)</span></span></code></pre></div>
<p><img src="multiRegionPlot_files/figure-html/aggPlot2-1.png" width="576"></p>
</div>
<div class="section level2">
<h2 id="visualizing-dname-and-sparse-signals">Visualizing DNAme and sparse signals<a class="anchor" aria-label="anchor" href="#visualizing-dname-and-sparse-signals"></a>
</h2>
<p>Nucleotide-resolution DNA methylation (as obtained from bisulfite
sequencing) signal differs from the signals used throughout this
vignette in that it is not continuous across the genome, but
specifically at C or CpG nucleotides which have a variable density
throughout the genome. As a consequence, it is likely that some of the
plotting bins do not contain a CpG, in which case they get assigned a
value of 0, even though they could be in a completely methylated region.
For this reason, it is advisable to smooth DNA methylation signals for
the purpose of visualization.</p>
<p>As an example, let’s look at the gene bodies of some active genes
from chr8 of the A549 cell lines:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"exampleDNAme"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">exampleDNAme</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## UnstitchedGPos object with 6 positions and 1 metadata column:</span></span>
<span><span class="co">##       seqnames       pos strand |     score</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;integer&gt;  &lt;Rle&gt; | &lt;integer&gt;</span></span>
<span><span class="co">##   [1]     chr9   4791172      + |        96</span></span>
<span><span class="co">##   [2]     chr9   4791391      + |        96</span></span>
<span><span class="co">##   [3]     chr9   4791398      + |       100</span></span>
<span><span class="co">##   [4]     chr9   4791422      + |        92</span></span>
<span><span class="co">##   [5]     chr9   4791437      + |        82</span></span>
<span><span class="co">##   [6]     chr9   4791627      + |        93</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 5 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<p>As is typical of DNAme data, the object is a GRanges object (or more
specifically a GPos object, since all ranges have a width of 1
nucleotide) with, in the score column, the percentage of DNA
methylation. Let’s see what happens if we plot a heatmap of this signal,
with and without smoothing (we use <code>type="scaled"</code> to scale
the gene bodies to the same size, since these can have very different
sizes) :</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">o1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/signal2Matrix.html">signal2Matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>noSmooth<span class="op">=</span><span class="va">exampleDNAme</span><span class="op">)</span>, <span class="va">geneBodies</span>, type<span class="op">=</span><span class="st">"scaled"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Computing signal from GRanges 'noSmooth'...</span></span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">o2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/signal2Matrix.html">signal2Matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>smoothed<span class="op">=</span><span class="va">exampleDNAme</span><span class="op">)</span>, <span class="va">geneBodies</span>, type<span class="op">=</span><span class="st">"scaled"</span>, </span>
<span>                    smooth<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Computing signal from GRanges 'smoothed'...</span></span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">o</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">o1</span>,<span class="va">o2</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">o</span>, scale_title<span class="op">=</span><span class="st">"%\nmethylation"</span>, axis_name<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TSS"</span>,<span class="st">"TES"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="multiRegionPlot_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>Both heatmaps show a very clear absence of DNA methylation at the
promoter of these genes (upstream of the TSS) an predominantly
methylated gene bodies. However they disagree substantially on the
methylation levels upstream the promoter and downstream the
transcription end sites (TES). This is because of the density of these
regions in (covered) CpG nucleotides. Since most genes are rather long,
most of the bins in the heatmap contain a CpG, leading to an actual
methylation signal. In the flanking regions, however, this is not
necessarily the case, and the non-smoothed heatmap does not distinguish
bins that are unmethylated form bins for which there is no information.
Instead, the smoothed heatmap on the right uses neighborhing bins to
estimate the methylation status of each bin, effectively filling out the
gaps. In doing so it provides the truthful representation, i.e. that the
regions downstream of the genes and upstream of the promoters are, most
of the time, as methylated as the gene bodies.</p>
<p>Smoothing is performed by <em><a href="https://bioconductor.org/packages/3.19/EnrichedHeatmap" class="external-link">EnrichedHeatmap</a></em>;
see <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html" class="external-link">?EnrichedHeatmap::normalizeToMatrix</a></code> for more
information/customization.</p>
<p><br><br></p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] ggplot2_3.5.1               epiwraps_0.99.95           </span></span>
<span><span class="co">##  [3] EnrichedHeatmap_1.34.0      ComplexHeatmap_2.20.0      </span></span>
<span><span class="co">##  [5] SummarizedExperiment_1.34.0 Biobase_2.64.0             </span></span>
<span><span class="co">##  [7] GenomicRanges_1.56.1        GenomeInfoDb_1.40.1        </span></span>
<span><span class="co">##  [9] IRanges_2.38.0              S4Vectors_0.42.0           </span></span>
<span><span class="co">## [11] BiocGenerics_0.50.0         MatrixGenerics_1.16.0      </span></span>
<span><span class="co">## [13] matrixStats_1.3.0           BiocStyle_2.32.1           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3       rstudioapi_0.16.0        jsonlite_1.8.8          </span></span>
<span><span class="co">##   [4] shape_1.4.6.1            magrittr_2.0.3           magick_2.8.3            </span></span>
<span><span class="co">##   [7] GenomicFeatures_1.56.0   farver_2.1.2             rmarkdown_2.27          </span></span>
<span><span class="co">##  [10] GlobalOptions_0.1.2      fs_1.6.4                 BiocIO_1.14.0           </span></span>
<span><span class="co">##  [13] zlibbioc_1.50.0          ragg_1.3.2               vctrs_0.6.5             </span></span>
<span><span class="co">##  [16] memoise_2.0.1            Rsamtools_2.20.0         RCurl_1.98-1.14         </span></span>
<span><span class="co">##  [19] base64enc_0.1-3          htmltools_0.5.8.1        S4Arrays_1.4.1          </span></span>
<span><span class="co">##  [22] progress_1.2.3           curl_5.2.1               SparseArray_1.4.8       </span></span>
<span><span class="co">##  [25] Formula_1.2-5            sass_0.4.9               bslib_0.7.0             </span></span>
<span><span class="co">##  [28] htmlwidgets_1.6.4        desc_1.4.3               plyr_1.8.9              </span></span>
<span><span class="co">##  [31] Gviz_1.48.0              httr2_1.0.1              cachem_1.1.0            </span></span>
<span><span class="co">##  [34] GenomicAlignments_1.40.0 lifecycle_1.0.4          iterators_1.0.14        </span></span>
<span><span class="co">##  [37] pkgconfig_2.0.3          Matrix_1.7-0             R6_2.5.1                </span></span>
<span><span class="co">##  [40] fastmap_1.2.0            GenomeInfoDbData_1.2.12  clue_0.3-65             </span></span>
<span><span class="co">##  [43] digest_0.6.36            colorspace_2.1-0         AnnotationDbi_1.66.0    </span></span>
<span><span class="co">##  [46] textshaping_0.4.0        Hmisc_5.1-3              RSQLite_2.3.7           </span></span>
<span><span class="co">##  [49] labeling_0.4.3           filelock_1.0.3           fansi_1.0.6             </span></span>
<span><span class="co">##  [52] httr_1.4.7               abind_1.4-5              compiler_4.4.1          </span></span>
<span><span class="co">##  [55] withr_3.0.0              bit64_4.0.5              doParallel_1.0.17       </span></span>
<span><span class="co">##  [58] backports_1.5.0          htmlTable_2.4.2          BiocParallel_1.38.0     </span></span>
<span><span class="co">##  [61] DBI_1.2.3                UpSetR_1.4.0             highr_0.11              </span></span>
<span><span class="co">##  [64] biomaRt_2.60.1           rappdirs_0.3.3           DelayedArray_0.30.1     </span></span>
<span><span class="co">##  [67] rjson_0.2.21             tools_4.4.1              foreign_0.8-86          </span></span>
<span><span class="co">##  [70] nnet_7.3-19              glue_1.7.0               restfulr_0.0.15         </span></span>
<span><span class="co">##  [73] checkmate_2.3.1          cluster_2.1.6            generics_0.1.3          </span></span>
<span><span class="co">##  [76] gtable_0.3.5             BSgenome_1.72.0          ensembldb_2.28.0        </span></span>
<span><span class="co">##  [79] data.table_1.15.4        hms_1.1.3                xml2_1.3.6              </span></span>
<span><span class="co">##  [82] utf8_1.2.4               XVector_0.44.0           foreach_1.5.2           </span></span>
<span><span class="co">##  [85] pillar_1.9.0             stringr_1.5.1            circlize_0.4.16         </span></span>
<span><span class="co">##  [88] dplyr_1.1.4              BiocFileCache_2.12.0     lattice_0.22-6          </span></span>
<span><span class="co">##  [91] deldir_2.0-4             rtracklayer_1.64.0       bit_4.0.5               </span></span>
<span><span class="co">##  [94] biovizBase_1.52.0        tidyselect_1.2.1         locfit_1.5-9.10         </span></span>
<span><span class="co">##  [97] pbapply_1.7-2            Biostrings_2.72.1        knitr_1.47              </span></span>
<span><span class="co">## [100] gridExtra_2.3            bookdown_0.40            ProtGenerics_1.36.0     </span></span>
<span><span class="co">## [103] xfun_0.45                stringi_1.8.4            UCSC.utils_1.0.0        </span></span>
<span><span class="co">## [106] lazyeval_0.2.2           yaml_2.3.8               evaluate_0.24.0         </span></span>
<span><span class="co">## [109] codetools_0.2-20         interp_1.1-6             GenomicFiles_1.40.0     </span></span>
<span><span class="co">## [112] tibble_3.2.1             BiocManager_1.30.23      cli_3.6.3               </span></span>
<span><span class="co">## [115] rpart_4.1.23             systemfonts_1.1.0        munsell_0.5.1           </span></span>
<span><span class="co">## [118] jquerylib_0.1.4          dichromat_2.0-0.1        Rcpp_1.0.12             </span></span>
<span><span class="co">## [121] dbplyr_2.5.0             png_0.1-8                XML_3.99-0.17           </span></span>
<span><span class="co">## [124] parallel_4.4.1           pkgdown_2.0.9            blob_1.2.4              </span></span>
<span><span class="co">## [127] prettyunits_1.2.0        jpeg_0.1-10              latticeExtra_0.6-30     </span></span>
<span><span class="co">## [130] AnnotationFilter_1.28.0  bitops_1.0-7             viridisLite_0.4.2       </span></span>
<span><span class="co">## [133] VariantAnnotation_1.50.0 scales_1.3.0             purrr_1.0.2             </span></span>
<span><span class="co">## [136] crayon_1.5.3             GetoptLong_1.0.5         rlang_1.1.4             </span></span>
<span><span class="co">## [139] cowplot_1.1.3            KEGGREST_1.44.1</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre-Luc Germain, Emanuel Sonder.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
