<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Normalizing genomic signals • epiwraps</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Normalizing genomic signals">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">epiwraps</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.107</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/bam2bw.html">Generating bigwig tracks with bam2bw</a></li>
    <li><a class="dropdown-item" href="../articles/misc.html">Miscellaneous epiwraps functions</a></li>
    <li><a class="dropdown-item" href="../articles/multiRegionPlot.html">Visualizing signals across many regions</a></li>
    <li><a class="dropdown-item" href="../articles/normalization.html">Normalizing genomic signals</a></li>
    <li><a class="dropdown-item" href="../articles/singleRegionPlot.html">Visualizing signals in a single region</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Normalizing genomic signals</h1>
                        <h4 data-toc-skip class="author">Pierre-Luc
Germain</h4>
            <address class="author_afil">
      Lab of Statistical Bioinformatics, University of Zürich;D-HEST
Institute for Neuroscience, ETH Zürich,
Switzerland<br><div class="d-none name"><code>normalization.Rmd</code></div>
    </address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>This vignette covers the functions for normalizing genomic
      signals. Since this is illustrated with visualization, it is
      recommended that you read the <code>bam2bw</code> and
      <code>multiRegionPlot</code> vignettes first.</p>
    </div>
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>epiwraps</code> includes two ways of calculating normalization
factors: either from the signal files (e.g. bam or bigwig files), which
is the most robust way and enables all options, or from an
<code>EnrichmentSE</code> object (see the
<a href="multiRegionPlot.html">multiRegionPlot vignette</a> for an intro
to such an object) or signal matrices. In both cases, the logic is the
same: we estimate normalization factors (mostly single linear scaling
factors, although some methods involve more complex normalization), and
then apply them to signals that were extracted using
<code><a href="../reference/signal2Matrix.html">signal2Matrix()</a></code>.</p>
<div class="section level3">
<h3 id="applying-normalization-factors-when-generating-the-bigwig-files">Applying normalization factors when generating the bigwig files<a class="anchor" aria-label="anchor" href="#applying-normalization-factors-when-generating-the-bigwig-files"></a>
</h3>
<p>It is possible to also directly use computed normalization factors
when creating bigwig files. By default, the
<a href="bam2bw.html"><code>bam2bw() function</code></a> scales using
library size, which can be disabled using <code>scaling=FALSE</code>.
However, it is also possible to pass the <code>scaling</code> argument a
manual scaling factor, as computed by the functions described here. In
this vignette, however, we will focus on normalizing signal
matrices.</p>
</div>
</div>
<div class="section level2">
<h2 id="obtaining-normalization-factors-for-a-set-of-signal-files">Obtaining normalization factors for a set of signal files<a class="anchor" aria-label="anchor" href="#obtaining-normalization-factors-for-a-set-of-signal-files"></a>
</h2>
<p>The <code><a href="../reference/getNormFactors.html">getNormFactors()</a></code> function can be used to estimate
normalization factors from either bam or bigwig files. The files cannot
be mixed (bam/bigwig), however, and it is important to note that
<em>normalization factors calculated on bam files cannot be applied to
data extracted from bigwig files, or vice versa</em>, because the bigwig
files are by default already normalized for library size. If needed,
however, <code><a href="../reference/getNormFactors.html">getNormFactors()</a></code> can be used to apply the same
method to both kind of files.</p>
<div class="section level3">
<h3 id="normalization-methods">Normalization methods<a class="anchor" aria-label="anchor" href="#normalization-methods"></a>
</h3>
<p>Simple library size normalization, as done by <code><a href="../reference/bam2bw.html">bam2bw()</a></code>,
is not always appropriate. The main reasons are 1) that different
samples/experiments can have a different signal-to-noise ratio, with the
result that more sequencing is needed to obtain a similar coverage of
enriched region; 2) that there might be global differences in the amount
of the signal of interest (e.g. more or less binding, globally, in one
cell type vs another); and 3) that there might be differences in
technical biases, such as GC content. For these reasons, different
normalization methods are needed according to circumstances and what
assumptions seem reasonable. Here is an overview of the normalization
methods currently implemented in <code>epiwraps</code> via the
<code><a href="../reference/getNormFactors.html">getNormFactors()</a></code> function:</p>
<ul>
<li>The ‘background’ or ‘SES’ normalization method (they are synonyms
here) assumes that the background noise should on average be the same
across experiments
(<a href="https://doi.org/10.1515/1544-6115.1750" class="external-link">Diaz et al., Stat Appl
Gen et Mol Biol, 2012</a>), an assumption that works well in practice
and is robust to global differences in the amount of signal when there
are not large differences in signal-to-noise ratio.</li>
<li>The ‘MAnorm’ approach (
<a href="https://doi.org/10.1186/gb-2012-13-3-r16" class="external-link">Shao et al., Genome
Biology 2012</a><a></a>) assumes that regions that are commonly enriched (i.e. 
common peaks) in two experiments should on average have the same signal
in the two experiments.</li>
<li>The ‘enriched’ approach assumes that enriched regions are on average
similarly enriched across samples. Contrarily to ‘MAnorm’, these regions
do not need to be in common across samples/experiments. This is not
robust to global differences.</li>
<li>The ‘top’ approach assumes that the maximum enrichment (after some
trimming) in peaks is the same across samples/experiments.</li>
<li>The ‘S3norm’ (<a href="https://doi.org/10.1093/nar/gkaa105" class="external-link">Xiang et
al., NAR 2020</a>) and ‘2cLinear’ methods try to normalize both
enrichment and background simultaneously. S3norm does this in a
log-linear fashion (as in the publication), while ‘2cLinear’ does it on
the original scale.</li>
</ul>
<p>The normalization factors can be computed using
<code><a href="../reference/getNormFactors.html">getNormFactors()</a></code> :</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ethz-ins.github.io/epiwraps/">epiwraps</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># we fetch the path to the example bigwig file:</span></span>
<span><span class="va">bwf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example_atac.bw"</span>, package<span class="op">=</span><span class="st">"epiwraps"</span><span class="op">)</span></span>
<span><span class="co"># we'll just double it to create a fake multi-sample dataset:</span></span>
<span><span class="va">bwfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>atac1<span class="op">=</span><span class="va">bwf</span>, atac2<span class="op">=</span><span class="va">bwf</span><span class="op">)</span></span>
<span><span class="va">nf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getNormFactors.html">getNormFactors</a></span><span class="op">(</span><span class="va">bwfiles</span>, method<span class="op">=</span><span class="st">"background"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Comparing coverage in random regions...</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nf</span></span></code></pre></div>
<pre><code><span><span class="co">## atac1 atac2 </span></span>
<span><span class="co">##     1     1</span></span></code></pre>
<p>In this case, since the files are identical, the factors are both
1.</p>
<p>Some normalization methods addtionally require peaks as input,
e.g.:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example_peaks.bed"</span>, package<span class="op">=</span><span class="st">"epiwraps"</span><span class="op">)</span></span>
<span><span class="va">nf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getNormFactors.html">getNormFactors</a></span><span class="op">(</span><span class="va">bwfiles</span>, peaks <span class="op">=</span> <span class="va">peaks</span>, method<span class="op">=</span><span class="st">"MAnorm"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Comparing coverage in peaks...</span></span></code></pre>
<p>(Note that MAnorm would normally require to have a list of peaks for
each sample/experiment).</p>
<p>Once computed, the normalization factors can be applied to an
<code>EnrichmentSE</code> object:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/signal2Matrix.html">signal2Matrix</a></span><span class="op">(</span><span class="va">bwfiles</span>, <span class="va">peaks</span>, extend<span class="op">=</span><span class="fl">1000L</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Reading /home/runner/work/_temp/Library/epiwraps/extdata/example_atac.bw</span></span>
<span><span class="co">## Reading /home/runner/work/_temp/Library/epiwraps/extdata/example_atac.bw</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renormalizeSignalMatrices.html">renormalizeSignalMatrices</a></span><span class="op">(</span><span class="va">sm</span>, scaleFactors<span class="op">=</span><span class="va">nf</span><span class="op">)</span></span>
<span><span class="va">sm</span></span></code></pre></div>
<pre><code><span><span class="co">## class: EnrichmentSE </span></span>
<span><span class="co">## 2 tracks across 150 regions</span></span>
<span><span class="co">## assays(2): normalized input</span></span>
<span><span class="co">## rownames(150): 1:195054101-195054250 1:133522798-133523047 ...</span></span>
<span><span class="co">##   1:22224734-22224983 1:90375438-90375787</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(2): atac1 atac2</span></span>
<span><span class="co">## colData names(0):</span></span>
<span><span class="co">## metadata(0):</span></span></code></pre>
<p>The object now has a new assay, called <code>normalized</code>, which
has been put in front and therefore will be used for most downstream
usages unless the uses specifies otherwise. Note that for any downstream
function it is however possible to specify which assay to use via the
<code>assay</code> argument.</p>
</div>
</div>
<div class="section level2">
<h2 id="obtaining-normalization-factors-from-the-signal-matrices-themselves">Obtaining normalization factors from the signal matrices
themselves<a class="anchor" aria-label="anchor" href="#obtaining-normalization-factors-from-the-signal-matrices-themselves"></a>
</h2>
<p>It is also possible to normalize the signal matrices using factors
derived from the matrices themselves, using the
<code>renormalizeSignalMatrices</code> function. Note that this is
provided as a ‘quick-and-dirty’ approach that does not have the
robustness of proper estimation methods. Specifically, beyond providing
manual scaling factors (e.g. computed using
<code>getNormFactors</code>), the function includes two methods :</p>
<ul>
<li>
<code>method="border"</code> works on the assumption that the
left/right borders of the matrices represent background signal which
should be equal across samples. As such, it can be seen as an
approximation of the aforementioned background normalization. However,
it will work only if 1) the left/right borders of the matrices are
sufficiently far from the signal (e.g. peaks) to be chiefly noise, and
(as with the main background normalization method itself) 2) the
signal-to-noise ratio is comparable across tracks/samples.</li>
<li>
<code>method="top"</code> instead works on the assumption that the
highest signal (after some eventual trimming of the extremes) should be
the same across tracks/samples.</li>
</ul>
<p>To illustrate these, we will first introduce some difference between
our two tracks using arbitrary factors:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renormalizeSignalMatrices.html">renormalizeSignalMatrices</a></span><span class="op">(</span><span class="va">sm</span>, scaleFactors<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">4</span><span class="op">)</span>, toAssay<span class="op">=</span><span class="st">"test"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">sm</span>, assay <span class="op">=</span> <span class="st">"test"</span><span class="op">)</span></span></code></pre></div>
<p><img src="normalization_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>Then we can normalize:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renormalizeSignalMatrices.html">renormalizeSignalMatrices</a></span><span class="op">(</span><span class="va">sm</span>, method<span class="op">=</span><span class="st">"top"</span>, fromAssay<span class="op">=</span><span class="st">"test"</span><span class="op">)</span></span>
<span><span class="co"># again this adds an assay to the object, which will be automatically used when plotting:</span></span>
<span><span class="fu"><a href="../reference/plotEnrichedHeatmaps.html">plotEnrichedHeatmaps</a></span><span class="op">(</span><span class="va">sm</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Using assay topNormalized</span></span></code></pre>
<p><img src="normalization_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>And we’ve recovered comparable signal across the two
tracks/samples.</p>
<p><br><br></p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.3 (2025-02-28)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] epiwraps_0.99.107           EnrichedHeatmap_1.36.0     </span></span>
<span><span class="co">##  [3] ComplexHeatmap_2.22.0       SummarizedExperiment_1.36.0</span></span>
<span><span class="co">##  [5] Biobase_2.66.0              GenomicRanges_1.58.0       </span></span>
<span><span class="co">##  [7] GenomeInfoDb_1.42.3         IRanges_2.40.1             </span></span>
<span><span class="co">##  [9] S4Vectors_0.44.0            BiocGenerics_0.52.0        </span></span>
<span><span class="co">## [11] MatrixGenerics_1.18.1       matrixStats_1.5.0          </span></span>
<span><span class="co">## [13] BiocStyle_2.34.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3       rstudioapi_0.17.1        jsonlite_1.9.0          </span></span>
<span><span class="co">##   [4] shape_1.4.6.1            magrittr_2.0.3           magick_2.8.5            </span></span>
<span><span class="co">##   [7] GenomicFeatures_1.58.0   rmarkdown_2.29           GlobalOptions_0.1.2     </span></span>
<span><span class="co">##  [10] fs_1.6.5                 BiocIO_1.16.0            zlibbioc_1.52.0         </span></span>
<span><span class="co">##  [13] ragg_1.3.3               vctrs_0.6.5              memoise_2.0.1           </span></span>
<span><span class="co">##  [16] Rsamtools_2.22.0         RCurl_1.98-1.16          base64enc_0.1-3         </span></span>
<span><span class="co">##  [19] htmltools_0.5.8.1        S4Arrays_1.6.0           progress_1.2.3          </span></span>
<span><span class="co">##  [22] curl_6.2.1               SparseArray_1.6.2        Formula_1.2-5           </span></span>
<span><span class="co">##  [25] sass_0.4.9               bslib_0.9.0              htmlwidgets_1.6.4       </span></span>
<span><span class="co">##  [28] desc_1.4.3               plyr_1.8.9               Gviz_1.50.0             </span></span>
<span><span class="co">##  [31] httr2_1.1.0              cachem_1.1.0             GenomicAlignments_1.42.0</span></span>
<span><span class="co">##  [34] lifecycle_1.0.4          iterators_1.0.14         pkgconfig_2.0.3         </span></span>
<span><span class="co">##  [37] Matrix_1.7-2             R6_2.6.1                 fastmap_1.2.0           </span></span>
<span><span class="co">##  [40] GenomeInfoDbData_1.2.13  clue_0.3-66              digest_0.6.37           </span></span>
<span><span class="co">##  [43] colorspace_2.1-1         AnnotationDbi_1.68.0     textshaping_1.0.0       </span></span>
<span><span class="co">##  [46] Hmisc_5.2-2              RSQLite_2.3.9            filelock_1.0.3          </span></span>
<span><span class="co">##  [49] httr_1.4.7               abind_1.4-8              compiler_4.4.3          </span></span>
<span><span class="co">##  [52] bit64_4.6.0-1            doParallel_1.0.17        backports_1.5.0         </span></span>
<span><span class="co">##  [55] htmlTable_2.4.3          BiocParallel_1.40.0      DBI_1.2.3               </span></span>
<span><span class="co">##  [58] UpSetR_1.4.0             biomaRt_2.62.1           rappdirs_0.3.3          </span></span>
<span><span class="co">##  [61] DelayedArray_0.32.0      rjson_0.2.23             tools_4.4.3             </span></span>
<span><span class="co">##  [64] foreign_0.8-88           nnet_7.3-20              glue_1.8.0              </span></span>
<span><span class="co">##  [67] restfulr_0.0.15          checkmate_2.3.2          cluster_2.1.8           </span></span>
<span><span class="co">##  [70] generics_0.1.3           gtable_0.3.6             BSgenome_1.74.0         </span></span>
<span><span class="co">##  [73] ensembldb_2.30.0         data.table_1.17.0        hms_1.1.3               </span></span>
<span><span class="co">##  [76] xml2_1.3.7               XVector_0.46.0           foreach_1.5.2           </span></span>
<span><span class="co">##  [79] pillar_1.10.1            stringr_1.5.1            limma_3.62.2            </span></span>
<span><span class="co">##  [82] circlize_0.4.16          dplyr_1.1.4              BiocFileCache_2.14.0    </span></span>
<span><span class="co">##  [85] lattice_0.22-6           deldir_2.0-4             rtracklayer_1.66.0      </span></span>
<span><span class="co">##  [88] bit_4.5.0.1              biovizBase_1.54.0        tidyselect_1.2.1        </span></span>
<span><span class="co">##  [91] locfit_1.5-9.11          pbapply_1.7-2            Biostrings_2.74.1       </span></span>
<span><span class="co">##  [94] knitr_1.49               gridExtra_2.3            bookdown_0.42           </span></span>
<span><span class="co">##  [97] ProtGenerics_1.38.0      edgeR_4.4.2              xfun_0.51               </span></span>
<span><span class="co">## [100] statmod_1.5.0            stringi_1.8.4            UCSC.utils_1.2.0        </span></span>
<span><span class="co">## [103] lazyeval_0.2.2           yaml_2.3.10              evaluate_1.0.3          </span></span>
<span><span class="co">## [106] codetools_0.2-20         interp_1.1-6             GenomicFiles_1.42.0     </span></span>
<span><span class="co">## [109] tibble_3.2.1             BiocManager_1.30.25      cli_3.6.4               </span></span>
<span><span class="co">## [112] rpart_4.1.24             systemfonts_1.2.1        munsell_0.5.1           </span></span>
<span><span class="co">## [115] jquerylib_0.1.4          dichromat_2.0-0.1        Rcpp_1.0.14             </span></span>
<span><span class="co">## [118] dbplyr_2.5.0             png_0.1-8                XML_3.99-0.18           </span></span>
<span><span class="co">## [121] parallel_4.4.3           pkgdown_2.1.1            ggplot2_3.5.1           </span></span>
<span><span class="co">## [124] blob_1.2.4               prettyunits_1.2.0        jpeg_0.1-10             </span></span>
<span><span class="co">## [127] latticeExtra_0.6-30      AnnotationFilter_1.30.0  bitops_1.0-9            </span></span>
<span><span class="co">## [130] viridisLite_0.4.2        VariantAnnotation_1.52.0 scales_1.3.0            </span></span>
<span><span class="co">## [133] crayon_1.5.3             GetoptLong_1.0.5         rlang_1.1.5             </span></span>
<span><span class="co">## [136] cowplot_1.1.3            KEGGREST_1.46.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pierre-Luc Germain, Emanuel Sonder.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
