<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="epiwraps">
<title>Visualizing signals in a single region • epiwraps</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Visualizing signals in a single region">
<meta property="og:description" content="epiwraps">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">epiwraps</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.84</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/bam2bw.html">Generating bigwig tracks with bam2bw</a>
    <a class="dropdown-item" href="../articles/misc.html">Miscellaneous epiwraps functions</a>
    <a class="dropdown-item" href="../articles/multiRegionPlot.html">Visualizing signals across many regions</a>
    <a class="dropdown-item" href="../articles/singleRegionPlot.html">Visualizing signals in a single region</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Visualizing signals in a single region</h1>
                        <h4 data-toc-skip class="author">Pierre-Luc
Germain</h4>
            <address class="author_afil">
      Lab of Statistical Bioinformatics, University of Zürich;D-HEST
Institute for Neuroscience, ETH Zürich,
Switzerland<br><div class="d-none name"><code>singleRegionPlot.Rmd</code></div>
    </address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>This vignette documents the use of the ‘plotSignalTracks’ to
      generate genome-browser-like plots of signals and annotations
      along genomic coordinates in a single given region. It is chiefly
      a wrapper around the ‘Gviz’ package.</p>
    </div>
    
<div class="section level2">
<h2 id="plotting-signals-in-a-region">Plotting signals in a region<a class="anchor" aria-label="anchor" href="#plotting-signals-in-a-region"></a>
</h2>
<p>The <code>plotSignalTracks</code> function is a wrapper around the
<em><a href="https://bioconductor.org/packages/3.18/Gviz" class="external-link">Gviz</a></em>
package, which plots one or more signals along genomic coordinates (in a
genome-browser like fashion). The function lacks the full flexibility of
the <em><a href="https://bioconductor.org/packages/3.18/Gviz" class="external-link">Gviz</a></em>
package, but presents a considerable simpler interface, with automatic
default parameters, etc. It has two essential arguments: a (named) list
of files whose signal to display (can be a mixture of bigwig, bam, or
bed-like files), and the region in which to display the signals (can be
given as a GRanges or as a string). The function then automatically
determines the relevant track type and setting from the file types.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ethz-ins.github.io/epiwraps/">epiwraps</a></span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get the path to an example bigwig file:</span></span>
<span><span class="va">bwf1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example_rna.bw"</span>, package<span class="op">=</span><span class="st">"epiwraps"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotSignalTracks.html">plotSignalTracks</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>RNA<span class="op">=</span><span class="va">bwf1</span><span class="op">)</span>, region<span class="op">=</span><span class="st">"8:22165140-22212326"</span>, genomeAxis<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="singleRegionPlot_files/figure-html/signal1-1.png" width="576"></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># we could plot multiple tracks as follows:</span></span>
<span><span class="fu"><a href="../reference/plotSignalTracks.html">plotSignalTracks</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>track1<span class="op">=</span><span class="va">bwf1</span>, track2<span class="op">=</span><span class="va">bwf1</span><span class="op">)</span>, region<span class="op">=</span><span class="st">"8:22165140-22212326"</span><span class="op">)</span></span></code></pre></div>
<p><img src="singleRegionPlot_files/figure-html/signal1-2.png" width="576"></p>
<p><code>GRanges</code> objects can also be plotted as annotation tracks
alongside other data:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myregions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"8"</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">22166000</span>,<span class="fl">22202300</span><span class="op">)</span>, width<span class="op">=</span><span class="fl">3000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotSignalTracks.html">plotSignalTracks</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>RNA<span class="op">=</span><span class="va">bwf1</span>, regions<span class="op">=</span><span class="va">myregions</span><span class="op">)</span>, region<span class="op">=</span><span class="st">"8:22165140-22212326"</span><span class="op">)</span></span></code></pre></div>
<p><img src="singleRegionPlot_files/figure-html/signal2-1.png" width="576"></p>
<p>Colors, track display types, and such parameters can either be set
for all tracks or for each individual track, for example:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myregions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"8"</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">22166000</span>,<span class="fl">22202300</span><span class="op">)</span>, width<span class="op">=</span><span class="fl">3000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotSignalTracks.html">plotSignalTracks</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>RNA<span class="op">=</span><span class="va">bwf1</span>, regions<span class="op">=</span><span class="va">myregions</span><span class="op">)</span>, colors<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"black"</span><span class="op">)</span>,</span>
<span>                 region<span class="op">=</span><span class="st">"8:22165140-22212326"</span><span class="op">)</span></span></code></pre></div>
<p><img src="singleRegionPlot_files/figure-html/signal3-1.png" width="576"></p>
<p>For bam files, we can also plot individual reads:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># we fetch an example bam file:</span></span>
<span><span class="va">bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"ex1.bam"</span>, package<span class="op">=</span><span class="st">"Rsamtools"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotSignalTracks.html">plotSignalTracks</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"my bam file"</span><span class="op">=</span><span class="va">bam</span><span class="op">)</span>, <span class="st">"seq1:1-1500"</span>, type<span class="op">=</span><span class="st">"alignments"</span><span class="op">)</span></span></code></pre></div>
<p><img src="singleRegionPlot_files/figure-html/signalAlignments-1.png" width="700"></p>
<div class="section level3">
<h3 id="merging-signal-from-different-tracks">Merging signal from different tracks<a class="anchor" aria-label="anchor" href="#merging-signal-from-different-tracks"></a>
</h3>
<p>In addition to being displayed one below the other, data tracks can
be combined in different ways. To do this, the tracks can simply be
given in a nested fashion:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSignalTracks.html">plotSignalTracks</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>track1<span class="op">=</span><span class="va">bwf1</span>, combined<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">bwf1</span>,<span class="va">bwf1</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                 region<span class="op">=</span><span class="st">"8:22165140-22212326"</span><span class="op">)</span></span></code></pre></div>
<p>In this example we are always using the same track, but the first
element (‘track1’) plots the track alone, while the second (‘combined’)
merges the two given tracks. By default, the mean is shown, but this can
be controlled through the <code>aggregation</code> argument. In addition
to usual operations, the tracks can be overlayed on top of one another
(<code>aggregation='overlay'</code>), or shown as a heatmap
(<code>aggregation='heatmap'</code>).</p>
</div>
<div class="section level3">
<h3 id="using-an-ensdb-object">Using an EnsDb object<a class="anchor" aria-label="anchor" href="#using-an-ensdb-object"></a>
</h3>
<p>If an <code>EnsDb</code> object is available (see the <em><a href="https://bioconductor.org/packages/3.18/ensembldb" class="external-link">ensembldb</a></em>
package for a description of the class and its methods, and the <em><a href="https://bioconductor.org/packages/3.18/AnnotationHub" class="external-link">AnnotationHub</a></em>
package for a convenient way of fetching such annotation objects), two
additional options are available: first, instead of specifying the
region as coordinates, one can specify a gene or transcript name, and
the corresponding region will be fetched. In addition, the genes or
transcripts can be displayed. For example:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># we fetch the GRCh38 Ensembl 103 annotation (this is not run in the vignette,</span></span>
<span><span class="co"># as it takes some time to download the annotation the first time is used):</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">AnnotationHub</span><span class="op">)</span></span>
<span><span class="va">ah</span> <span class="op">&lt;-</span> <span class="fu">AnnotationHub</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">ensdb</span> <span class="op">&lt;-</span> <span class="va">ah</span><span class="op">[[</span><span class="st">"AH89426"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co"># we plot our previous RNA bigwig file, around the BMP1 locus:</span></span>
<span><span class="fu"><a href="../reference/plotSignalTracks.html">plotSignalTracks</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>coverage<span class="op">=</span><span class="va">bwf1</span><span class="op">)</span>, region<span class="op">=</span><span class="st">"BMP1"</span>, ensdb<span class="op">=</span><span class="va">ensdb</span>, </span>
<span>                 transcripts<span class="op">=</span><span class="st">"full"</span><span class="op">)</span></span></code></pre></div>
<p><img src="TracksWithTranscripts.png"></p>
<p>Now we can see that the coverage is nicely restricted to exons, and
that some transcripts/exons are not expressed as highly as others. The
transcripts could also have been collapsed into a gene model using
<code>transcripts="collapsed"</code> (the default).</p>
<p>To display only the gene track, the first argument can simply be
omitted.</p>
</div>
<div class="section level3">
<h3 id="further-track-customization">Further track customization<a class="anchor" aria-label="anchor" href="#further-track-customization"></a>
</h3>
<p>In addition to the <code>colors</code> and <code>type</code> argument
(and a number of others), which can customize the appearance of tracks,
any additional parameters supported by the respective <em><a href="https://bioconductor.org/packages/3.18/Gviz" class="external-link">Gviz</a></em>
function can be passed through the <code>genes.params</code> (for Gviz’s
<code>GeneRegionTrack</code>), <code>align.params</code> (for Gviz’s
<code>AlignmentsTrack</code>, when plotting individual reads), or
<code>tracks.params</code> (for any other Gviz
<code>DataTrack</code>).</p>
<p>For example, if you wish to manually set the same y-axis range for
all data tracks, this can be done with:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSignalTracks.html">plotSignalTracks</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>track1<span class="op">=</span><span class="va">bwf1</span>, track2<span class="op">=</span><span class="va">bwf1</span><span class="op">)</span>, region<span class="op">=</span><span class="st">"8:22165140-22212326"</span>,</span>
<span>                 tracks.params<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">200</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="singleRegionPlot_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<p>Also, in addition to passing filepaths or <code>GRanges</code>, any
Gviz track(s) can be passed (i.e. objects inheriting the
<code>GdObject</code> class) can be passed, enabling full track
customization when needed.</p>
<p><br><br></p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.3 (2024-02-29)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] epiwraps_0.99.84            EnrichedHeatmap_1.32.0     </span></span>
<span><span class="co">##  [3] ComplexHeatmap_2.18.0       SummarizedExperiment_1.32.0</span></span>
<span><span class="co">##  [5] Biobase_2.62.0              GenomicRanges_1.54.1       </span></span>
<span><span class="co">##  [7] GenomeInfoDb_1.38.8         IRanges_2.36.0             </span></span>
<span><span class="co">##  [9] S4Vectors_0.40.2            BiocGenerics_0.48.1        </span></span>
<span><span class="co">## [11] MatrixGenerics_1.14.0       matrixStats_1.3.0          </span></span>
<span><span class="co">## [13] BiocStyle_2.30.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3       rstudioapi_0.16.0        jsonlite_1.8.8          </span></span>
<span><span class="co">##   [4] shape_1.4.6.1            magrittr_2.0.3           GenomicFeatures_1.54.4  </span></span>
<span><span class="co">##   [7] rmarkdown_2.26           GlobalOptions_0.1.2      fs_1.6.3                </span></span>
<span><span class="co">##  [10] BiocIO_1.12.0            zlibbioc_1.48.2          ragg_1.3.0              </span></span>
<span><span class="co">##  [13] vctrs_0.6.5              memoise_2.0.1            Rsamtools_2.18.0        </span></span>
<span><span class="co">##  [16] RCurl_1.98-1.14          base64enc_0.1-3          htmltools_0.5.8.1       </span></span>
<span><span class="co">##  [19] S4Arrays_1.2.1           progress_1.2.3           curl_5.2.1              </span></span>
<span><span class="co">##  [22] SparseArray_1.2.4        Formula_1.2-5            sass_0.4.9              </span></span>
<span><span class="co">##  [25] bslib_0.7.0              htmlwidgets_1.6.4        desc_1.4.3              </span></span>
<span><span class="co">##  [28] plyr_1.8.9               Gviz_1.46.1              cachem_1.0.8            </span></span>
<span><span class="co">##  [31] GenomicAlignments_1.38.2 lifecycle_1.0.4          iterators_1.0.14        </span></span>
<span><span class="co">##  [34] pkgconfig_2.0.3          Matrix_1.6-5             R6_2.5.1                </span></span>
<span><span class="co">##  [37] fastmap_1.1.1            GenomeInfoDbData_1.2.11  clue_0.3-65             </span></span>
<span><span class="co">##  [40] digest_0.6.35            colorspace_2.1-0         AnnotationDbi_1.64.1    </span></span>
<span><span class="co">##  [43] textshaping_0.3.7        Hmisc_5.1-2              RSQLite_2.3.6           </span></span>
<span><span class="co">##  [46] filelock_1.0.3           fansi_1.0.6              httr_1.4.7              </span></span>
<span><span class="co">##  [49] abind_1.4-5              compiler_4.3.3           bit64_4.0.5             </span></span>
<span><span class="co">##  [52] doParallel_1.0.17        backports_1.4.1          htmlTable_2.4.2         </span></span>
<span><span class="co">##  [55] BiocParallel_1.36.0      DBI_1.2.2                UpSetR_1.4.0            </span></span>
<span><span class="co">##  [58] highr_0.10               biomaRt_2.58.2           rappdirs_0.3.3          </span></span>
<span><span class="co">##  [61] DelayedArray_0.28.0      rjson_0.2.21             tools_4.3.3             </span></span>
<span><span class="co">##  [64] foreign_0.8-86           nnet_7.3-19              glue_1.7.0              </span></span>
<span><span class="co">##  [67] restfulr_0.0.15          checkmate_2.3.1          cluster_2.1.6           </span></span>
<span><span class="co">##  [70] generics_0.1.3           gtable_0.3.4             BSgenome_1.70.2         </span></span>
<span><span class="co">##  [73] ensembldb_2.26.0         data.table_1.15.4        hms_1.1.3               </span></span>
<span><span class="co">##  [76] xml2_1.3.6               utf8_1.2.4               XVector_0.42.0          </span></span>
<span><span class="co">##  [79] foreach_1.5.2            pillar_1.9.0             stringr_1.5.1           </span></span>
<span><span class="co">##  [82] circlize_0.4.16          dplyr_1.1.4              BiocFileCache_2.10.2    </span></span>
<span><span class="co">##  [85] lattice_0.22-5           deldir_2.0-4             rtracklayer_1.62.0      </span></span>
<span><span class="co">##  [88] bit_4.0.5                biovizBase_1.50.0        tidyselect_1.2.1        </span></span>
<span><span class="co">##  [91] locfit_1.5-9.9           pbapply_1.7-2            Biostrings_2.70.3       </span></span>
<span><span class="co">##  [94] knitr_1.46               gridExtra_2.3            bookdown_0.39           </span></span>
<span><span class="co">##  [97] ProtGenerics_1.34.0      xfun_0.43                stringi_1.8.3           </span></span>
<span><span class="co">## [100] lazyeval_0.2.2           yaml_2.3.8               evaluate_0.23           </span></span>
<span><span class="co">## [103] codetools_0.2-19         interp_1.1-6             GenomicFiles_1.38.0     </span></span>
<span><span class="co">## [106] tibble_3.2.1             BiocManager_1.30.22      cli_3.6.2               </span></span>
<span><span class="co">## [109] rpart_4.1.23             systemfonts_1.0.6        munsell_0.5.1           </span></span>
<span><span class="co">## [112] jquerylib_0.1.4          dichromat_2.0-0.1        Rcpp_1.0.12             </span></span>
<span><span class="co">## [115] dbplyr_2.5.0             png_0.1-8                XML_3.99-0.16.1         </span></span>
<span><span class="co">## [118] parallel_4.3.3           pkgdown_2.0.8            ggplot2_3.5.0           </span></span>
<span><span class="co">## [121] blob_1.2.4               prettyunits_1.2.0        jpeg_0.1-10             </span></span>
<span><span class="co">## [124] latticeExtra_0.6-30      AnnotationFilter_1.26.0  bitops_1.0-7            </span></span>
<span><span class="co">## [127] viridisLite_0.4.2        VariantAnnotation_1.48.1 scales_1.3.0            </span></span>
<span><span class="co">## [130] purrr_1.0.2              crayon_1.5.2             GetoptLong_1.0.5        </span></span>
<span><span class="co">## [133] rlang_1.1.3              cowplot_1.1.3            KEGGREST_1.42.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Pierre-Luc Germain, Emanuel Sonder.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.8.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
